// aws-cdk-lib/aws-rds/test/database-secretmanager.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const assertions_1 = require("../../assertions");
const ec2 = require("../../aws-ec2");
const secretsmanager = require("../../aws-secretsmanager");
const cdk = require("../../core");
const lib_1 = require("../lib");
describe('database secret manager', () => {
    test('can create a Serverless Cluster using an existing secret from secretmanager', () => {
        // GIVEN
        const stack = testStack();
        const vpc = new ec2.Vpc(stack, 'VPC');
        const existingSecret = secretsmanager.Secret.fromSecretNameV2(stack, 'DBSecret', 'myDBLoginInfo');
        // WHEN
        new lib_1.ServerlessCluster(stack, 'ServerlessDatabase', {
            engine: lib_1.DatabaseClusterEngine.AURORA_POSTGRESQL,
            vpc,
            credentials: lib_1.Credentials.fromSecret(existingSecret),
            parameterGroup: lib_1.ParameterGroup.fromParameterGroupName(stack, 'ParameterGroup', 'default.aurora-postgresql10'),
        });
        // THEN
        assertions_1.Template.fromStack(stack).hasResource('AWS::RDS::DBCluster', {
            Properties: {
                Engine: 'aurora-postgresql',
                DBClusterParameterGroupName: 'default.aurora-postgresql10',
                DBSubnetGroupName: {
                    Ref: 'ServerlessDatabaseSubnets5643CD76',
                },
                EngineMode: 'serverless',
                MasterUsername: {
                    'Fn::Join': [
                        '',
                        [
                            '{{resolve:secretsmanager:arn:',
                            {
                                Ref: 'AWS::Partition',
                            },
                            ':secretsmanager:us-test-1:12345:secret:myDBLoginInfo:SecretString:username::}}',
                        ],
                    ],
                },
                MasterUserPassword: {
                    'Fn::Join': [
                        '',
                        [
                            '{{resolve:secretsmanager:arn:',
                            {
                                Ref: 'AWS::Partition',
                            },
                            ':secretsmanager:us-test-1:12345:secret:myDBLoginInfo:SecretString:password::}}',
                        ],
                    ],
                },
                StorageEncrypted: true,
                VpcSecurityGroupIds: [
                    {
                        'Fn::GetAtt': [
                            'ServerlessDatabaseSecurityGroupB00D8C0F',
                            'GroupId',
                        ],
                    },
                ],
            },
            DeletionPolicy: 'Snapshot',
            UpdateReplacePolicy: 'Snapshot',
        });
    });
});
function testStack() {
    const stack = new cdk.Stack(undefined, undefined, { env: { account: '12345', region: 'us-test-1' } });
    stack.node.setContext('availability-zones:12345:us-test-1', ['us-test-1a', 'us-test-1b']);
    return stack;
}
