// aws-cdk-lib/aws-rds/test/serverless-cluster-from-snapshot.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const assertions_1 = require("../../assertions");
const ec2 = require("../../aws-ec2");
const kms = require("../../aws-kms");
const cdk = require("../../core");
const lib_1 = require("../lib");
describe('serverless cluster from snapshot', () => {
    test('create a serverless cluster from a snapshot', () => {
        const stack = testStack();
        const vpc = new ec2.Vpc(stack, 'VPC');
        // WHEN
        new lib_1.ServerlessClusterFromSnapshot(stack, 'ServerlessDatabase', {
            engine: lib_1.DatabaseClusterEngine.AURORA_MYSQL,
            vpc,
            snapshotIdentifier: 'my-snapshot',
        });
        // THEN
        assertions_1.Template.fromStack(stack).hasResource('AWS::RDS::DBCluster', {
            Properties: {
                Engine: 'aurora-mysql',
                CopyTagsToSnapshot: true,
                DBClusterParameterGroupName: 'default.aurora-mysql5.7',
                DBSubnetGroupName: {
                    Ref: 'ServerlessDatabaseSubnets5643CD76',
                },
                EngineMode: 'serverless',
                SnapshotIdentifier: 'my-snapshot',
                StorageEncrypted: true,
                VpcSecurityGroupIds: [
                    {
                        'Fn::GetAtt': [
                            'ServerlessDatabaseSecurityGroupB00D8C0F',
                            'GroupId',
                        ],
                    },
                ],
            },
            DeletionPolicy: 'Snapshot',
            UpdateReplacePolicy: 'Snapshot',
        });
    });
    test('can generate a new snapshot password', () => {
        const stack = testStack();
        const vpc = new ec2.Vpc(stack, 'VPC');
        // WHEN
        new lib_1.ServerlessClusterFromSnapshot(stack, 'ServerlessDatabase', {
            engine: lib_1.DatabaseClusterEngine.AURORA_MYSQL,
            vpc,
            snapshotIdentifier: 'mySnapshot',
            credentials: lib_1.SnapshotCredentials.fromGeneratedSecret('admin', {
                excludeCharacters: '"@/\\',
            }),
        });
        // THEN
        assertions_1.Template.fromStack(stack).hasResourceProperties('AWS::RDS::DBCluster', {
            MasterUsername: assertions_1.Match.absent(),
            MasterUserPassword: {
                'Fn::Join': ['', [
                        '{{resolve:secretsmanager:',
                        { Ref: 'ServerlessDatabaseSecret813910E98ee0a797cad8a68dbeb85f8698cdb5bb' },
                        ':SecretString:password::}}',
                    ]],
            },
        });
        assertions_1.Template.fromStack(stack).hasResourceProperties('AWS::SecretsManager::Secret', {
            Description: {
                'Fn::Join': ['', ['Generated by the CDK for stack: ', { Ref: 'AWS::StackName' }]],
            },
            GenerateSecretString: {
                ExcludeCharacters: '\"@/\\',
                GenerateStringKey: 'password',
                PasswordLength: 30,
                SecretStringTemplate: '{"username":"admin"}',
            },
        });
    });
    test('fromGeneratedSecret with replica regions', () => {
        const stack = testStack();
        const vpc = new ec2.Vpc(stack, 'VPC');
        // WHEN
        new lib_1.ServerlessClusterFromSnapshot(stack, 'ServerlessDatabase', {
            engine: lib_1.DatabaseClusterEngine.AURORA_MYSQL,
            vpc,
            snapshotIdentifier: 'mySnapshot',
            credentials: lib_1.SnapshotCredentials.fromGeneratedSecret('admin', {
                replicaRegions: [{ region: 'eu-west-1' }],
            }),
        });
        // THEN
        assertions_1.Template.fromStack(stack).hasResourceProperties('AWS::SecretsManager::Secret', {
            ReplicaRegions: [
                {
                    Region: 'eu-west-1',
                },
            ],
        });
    });
    test('throws if generating a new password without a username', () => {
        const stack = testStack();
        const vpc = new ec2.Vpc(stack, 'VPC');
        // WHEN
        expect(() => new lib_1.ServerlessClusterFromSnapshot(stack, 'ServerlessDatabase', {
            engine: lib_1.DatabaseClusterEngine.AURORA_MYSQL,
            vpc,
            snapshotIdentifier: 'mySnapshot',
            credentials: { generatePassword: true },
        })).toThrow(/`credentials` `username` must be specified when `generatePassword` is set to true/);
    });
    test('can set a new snapshot password from an existing SecretValue', () => {
        const stack = testStack();
        const vpc = new ec2.Vpc(stack, 'VPC');
        // WHEN
        new lib_1.ServerlessClusterFromSnapshot(stack, 'ServerlessDatabase', {
            engine: lib_1.DatabaseClusterEngine.AURORA_MYSQL,
            vpc,
            snapshotIdentifier: 'mySnapshot',
            credentials: lib_1.SnapshotCredentials.fromPassword(cdk.SecretValue.unsafePlainText('mysecretpassword')),
        });
        // THEN
        assertions_1.Template.fromStack(stack).hasResourceProperties('AWS::RDS::DBCluster', {
            MasterUsername: assertions_1.Match.absent(),
            MasterUserPassword: 'mysecretpassword',
        });
    });
    test('can set a new snapshot password from an existing Secret', () => {
        const stack = testStack();
        const vpc = new ec2.Vpc(stack, 'VPC');
        // WHEN
        const secret = new lib_1.DatabaseSecret(stack, 'DBSecret', {
            username: 'admin',
            encryptionKey: new kms.Key(stack, 'PasswordKey'),
        });
        new lib_1.ServerlessClusterFromSnapshot(stack, 'ServerlessDatabase', {
            engine: lib_1.DatabaseClusterEngine.AURORA_MYSQL,
            vpc,
            snapshotIdentifier: 'mySnapshot',
            credentials: lib_1.SnapshotCredentials.fromSecret(secret),
        });
        // THEN
        assertions_1.Template.fromStack(stack).hasResourceProperties('AWS::RDS::DBCluster', {
            MasterUsername: assertions_1.Match.absent(),
            MasterUserPassword: {
                'Fn::Join': ['', ['{{resolve:secretsmanager:', { Ref: 'DBSecretD58955BC' }, ':SecretString:password::}}']],
            },
        });
    });
});
function testStack() {
    const stack = new cdk.Stack(undefined, undefined, { env: { account: '12345', region: 'us-test-1' } });
    stack.node.setContext('availability-zones:12345:us-test-1', ['us-test-1a', 'us-test-1b']);
    return stack;
}
