// aws-cdk-lib/aws-rds/test/subnet-group.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const assertions_1 = require("../../assertions");
const ec2 = require("../../aws-ec2");
const cdk = require("../../core");
const rds = require("../lib");
let stack;
let vpc;
describe('subnet group', () => {
    beforeEach(() => {
        stack = new cdk.Stack();
        vpc = new ec2.Vpc(stack, 'VPC');
    });
    test('creates a subnet group from minimal properties', () => {
        new rds.SubnetGroup(stack, 'Group', {
            description: 'MyGroup',
            vpc,
        });
        assertions_1.Template.fromStack(stack).hasResourceProperties('AWS::RDS::DBSubnetGroup', {
            DBSubnetGroupDescription: 'MyGroup',
            SubnetIds: [
                { Ref: 'VPCPrivateSubnet1Subnet8BCA10E0' },
                { Ref: 'VPCPrivateSubnet2SubnetCFCDAA7A' },
            ],
        });
    });
    test('creates a subnet group from all properties', () => {
        new rds.SubnetGroup(stack, 'Group', {
            description: 'My Shared Group',
            subnetGroupName: 'SharedGroup',
            vpc,
            vpcSubnets: { subnetType: ec2.SubnetType.PRIVATE_WITH_EGRESS },
        });
        assertions_1.Template.fromStack(stack).hasResourceProperties('AWS::RDS::DBSubnetGroup', {
            DBSubnetGroupDescription: 'My Shared Group',
            DBSubnetGroupName: 'sharedgroup',
            SubnetIds: [
                { Ref: 'VPCPrivateSubnet1Subnet8BCA10E0' },
                { Ref: 'VPCPrivateSubnet2SubnetCFCDAA7A' },
            ],
        });
    });
    test('correctly creates a subnet group with a deploy-time value for its name', () => {
        const parameter = new cdk.CfnParameter(stack, 'Parameter');
        new rds.SubnetGroup(stack, 'Group', {
            description: 'My Shared Group',
            subnetGroupName: parameter.valueAsString,
            vpc,
            vpcSubnets: { subnetType: ec2.SubnetType.PRIVATE_WITH_EGRESS },
        });
        assertions_1.Template.fromStack(stack).hasResourceProperties('AWS::RDS::DBSubnetGroup', {
            DBSubnetGroupName: {
                Ref: 'Parameter',
            },
        });
    });
    describe('subnet selection', () => {
        test('defaults to private subnets', () => {
            new rds.SubnetGroup(stack, 'Group', {
                description: 'MyGroup',
                vpc,
            });
            assertions_1.Template.fromStack(stack).hasResourceProperties('AWS::RDS::DBSubnetGroup', {
                DBSubnetGroupDescription: 'MyGroup',
                SubnetIds: [
                    { Ref: 'VPCPrivateSubnet1Subnet8BCA10E0' },
                    { Ref: 'VPCPrivateSubnet2SubnetCFCDAA7A' },
                ],
            });
        });
        test('can specify subnet type', () => {
            new rds.SubnetGroup(stack, 'Group', {
                description: 'MyGroup',
                vpc,
                vpcSubnets: { subnetType: ec2.SubnetType.PUBLIC },
            });
            assertions_1.Template.fromStack(stack).hasResourceProperties('AWS::RDS::DBSubnetGroup', {
                DBSubnetGroupDescription: 'MyGroup',
                SubnetIds: [
                    { Ref: 'VPCPublicSubnet1SubnetB4246D30' },
                    { Ref: 'VPCPublicSubnet2Subnet74179F39' },
                ],
            });
        });
    });
    test('import group by name', () => {
        const subnetGroup = rds.SubnetGroup.fromSubnetGroupName(stack, 'Group', 'my-subnet-group');
        expect(subnetGroup.subnetGroupName).toEqual('my-subnet-group');
    });
});
