// aws-cdk-lib/pipelines/test/testhelpers/compliance.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.behavior = void 0;
const cdk_build_tools_1 = require("@aws-cdk/cdk-build-tools");
// eslint-disable-next-line jest/no-export
function behavior(name, cb) {
    // Since the goal of the compliance test suites is to compare modern and legacy (i.e. deprecated) APIs,
    // use `describeDeprecated()` block here since usage of the legacy API is inevitable.
    (0, cdk_build_tools_1.describeDeprecated)(name, () => {
        const unwritten = new Set(['modern', 'legacy']);
        function scratchOff(flavor) {
            if (!unwritten.has(flavor)) {
                throw new Error(`Already had test for ${flavor}. Use .additional() to add more tests.`);
            }
            unwritten.delete(flavor);
        }
        cb({
            legacy: (testFn) => {
                scratchOff('legacy');
                test('legacy', testFn);
            },
            modern: (testFn) => {
                scratchOff('modern');
                test('modern', testFn);
            },
            additional: test,
            doesNotApply: {
                modern: (reason) => {
                    scratchOff('modern');
                    if (reason != null) {
                        // eslint-disable-next-line jest/no-disabled-tests
                        test.skip(`modern - ${reason}`, () => { });
                    }
                },
                legacy: (reason) => {
                    scratchOff('legacy');
                    if (reason != null) {
                        // eslint-disable-next-line jest/no-disabled-tests
                        test.skip(`legacy - ${reason}`, () => { });
                    }
                },
            },
        });
        for (const missing of unwritten) {
            test.todo(missing);
        }
    });
}
exports.behavior = behavior;
