// aws-cdk-lib/pipelines/test/testhelpers/legacy-pipeline.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TestGitHubAction = exports.LegacyTestGitHubNpmPipeline = void 0;
const codepipeline = require("../../../aws-codepipeline");
const codepipeline_actions = require("../../../aws-codepipeline-actions");
const core_1 = require("../../../core");
const cdkp = require("../../lib");
class LegacyTestGitHubNpmPipeline extends cdkp.CdkPipeline {
    constructor(scope, id, props) {
        const sourceArtifact = props?.sourceArtifact ?? new codepipeline.Artifact();
        const cloudAssemblyArtifact = props?.cloudAssemblyArtifact ?? new codepipeline.Artifact();
        super(scope, id, {
            sourceAction: new TestGitHubAction(sourceArtifact),
            synthAction: cdkp.SimpleSynthAction.standardNpmSynth({
                sourceArtifact,
                cloudAssemblyArtifact,
                ...props?.npmSynthOptions,
            }),
            cloudAssemblyArtifact,
            ...props,
        });
        this.sourceArtifact = sourceArtifact;
        this.cloudAssemblyArtifact = cloudAssemblyArtifact;
    }
}
exports.LegacyTestGitHubNpmPipeline = LegacyTestGitHubNpmPipeline;
class TestGitHubAction extends codepipeline_actions.GitHubSourceAction {
    constructor(sourceArtifact) {
        super({
            actionName: 'GitHub',
            output: sourceArtifact,
            oauthToken: core_1.SecretValue.unsafePlainText('$3kr1t'),
            owner: 'test',
            repo: 'test',
            trigger: codepipeline_actions.GitHubTrigger.POLL,
        });
    }
}
exports.TestGitHubAction = TestGitHubAction;
