// aws-cdk-lib/pipelines/test/testhelpers/matchers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.stringNoLongerThan = exports.sortByRunOrder = exports.stringLike = void 0;
const assertions_1 = require("../../../assertions");
function stringLike(pattern) {
    return new StringLike(pattern);
}
exports.stringLike = stringLike;
function sortByRunOrder(pattern) {
    return new Sorter('SortByRunOrder', pattern, (a, b) => {
        if (a.RunOrder !== b.RunOrder) {
            return a.RunOrder - b.RunOrder;
        }
        return a.Name.localeCompare(b.Name);
    });
}
exports.sortByRunOrder = sortByRunOrder;
function stringNoLongerThan(max) {
    return new StringLengthMatcher(max);
}
exports.stringNoLongerThan = stringNoLongerThan;
class StringLike extends assertions_1.Matcher {
    constructor(pattern) {
        super();
        this.pattern = pattern;
        this.name = 'StringLike';
    }
    test(actual) {
        if (typeof (actual) !== 'string') {
            throw new Error(`Expected string but found ${typeof (actual)} ${JSON.stringify(actual)}`);
        }
        const re = new RegExp(`^${this.pattern.split('*').map(escapeRegex).join('.*')}$`);
        const result = new assertions_1.MatchResult(actual);
        if (!re.test(actual)) {
            result.push(this, [], `Looking for string with pattern "${this.pattern}" but found "${actual}"`);
        }
        return result;
        function escapeRegex(s) {
            return s.replace(/[.*+?^${}()|[\]\\]/g, '\\$&'); // $& means the whole matched string
        }
    }
}
class Sorter extends assertions_1.Matcher {
    constructor(name, pattern, compareFn) {
        super();
        this.name = name;
        this.pattern = pattern;
        this.compareFn = compareFn;
    }
    test(actual) {
        const result = new assertions_1.MatchResult(actual);
        if (!Array.isArray(actual)) {
            result.push(this, [], `Expected an Array, but got '${typeof actual}'`);
            return result;
        }
        const copy = actual.slice();
        copy.sort(this.compareFn);
        const matcher = assertions_1.Matcher.isMatcher(this.pattern) ? this.pattern : assertions_1.Match.exact(this.pattern);
        return matcher.test(copy);
    }
}
class StringLengthMatcher extends assertions_1.Matcher {
    constructor(length) {
        super();
        this.length = length;
        this.name = 'StringLength';
    }
    test(actual) {
        const result = new assertions_1.MatchResult(actual);
        if (typeof actual !== 'string') {
            result.push(this, [], `Expected a string, but got '${typeof actual}'`);
        }
        if (actual.length > this.length) {
            result.push(this, [], `String is ${actual.length} characters long. Expected at most ${this.length} characters`);
        }
        return result;
    }
}
