// aws-cdk-lib/pipelines/test/codepipeline/codepipeline-sources.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const assertions_1 = require("../../../assertions");
const ccommit = require("../../../aws-codecommit");
const aws_codepipeline_actions_1 = require("../../../aws-codepipeline-actions");
const ecr = require("../../../aws-ecr");
const aws_iam_1 = require("../../../aws-iam");
const s3 = require("../../../aws-s3");
const core_1 = require("../../../core");
const cdkp = require("../../lib");
const testhelpers_1 = require("../testhelpers");
let app;
let pipelineStack;
beforeEach(() => {
    app = new testhelpers_1.TestApp();
    pipelineStack = new core_1.Stack(app, 'PipelineStack', { env: testhelpers_1.PIPELINE_ENV });
});
afterEach(() => {
    app.cleanup();
});
test('CodeCommit source handles tokenized names correctly', () => {
    const repo = new ccommit.Repository(pipelineStack, 'Repo', {
        repositoryName: 'MyRepo',
    });
    new testhelpers_1.ModernTestGitHubNpmPipeline(pipelineStack, 'Pipeline', {
        input: cdkp.CodePipelineSource.codeCommit(repo, 'main'),
    });
    assertions_1.Template.fromStack(pipelineStack).hasResourceProperties('AWS::CodePipeline::Pipeline', {
        Stages: assertions_1.Match.arrayWith([{
                Name: 'Source',
                Actions: [
                    assertions_1.Match.objectLike({
                        Configuration: assertions_1.Match.objectLike({
                            RepositoryName: { 'Fn::GetAtt': [assertions_1.Match.anyValue(), 'Name'] },
                        }),
                        Name: { 'Fn::GetAtt': [assertions_1.Match.anyValue(), 'Name'] },
                    }),
                ],
            }]),
    });
});
test('CodeCommit source honors all valid properties', () => {
    const repo = new ccommit.Repository(pipelineStack, 'Repo', {
        repositoryName: 'MyRepo',
    });
    new testhelpers_1.ModernTestGitHubNpmPipeline(pipelineStack, 'Pipeline', {
        input: cdkp.CodePipelineSource.codeCommit(repo, 'main', {
            codeBuildCloneOutput: true,
            trigger: aws_codepipeline_actions_1.CodeCommitTrigger.POLL,
            eventRole: new aws_iam_1.Role(pipelineStack, 'role', {
                assumedBy: new aws_iam_1.AnyPrincipal(),
                roleName: 'some-role',
            }),
        }),
    });
    assertions_1.Template.fromStack(pipelineStack).hasResourceProperties('AWS::CodePipeline::Pipeline', {
        Stages: assertions_1.Match.arrayWith([{
                Name: 'Source',
                Actions: [
                    assertions_1.Match.objectLike({
                        Configuration: assertions_1.Match.objectLike({
                            BranchName: 'main',
                            PollForSourceChanges: true,
                            OutputArtifactFormat: 'CODEBUILD_CLONE_REF',
                        }),
                        RoleArn: { 'Fn::GetAtt': [assertions_1.Match.anyValue(), 'Arn'] },
                    }),
                ],
            }]),
    });
});
test('S3 source handles tokenized names correctly', () => {
    const bucket = new s3.Bucket(pipelineStack, 'Bucket');
    new testhelpers_1.ModernTestGitHubNpmPipeline(pipelineStack, 'Pipeline', {
        input: cdkp.CodePipelineSource.s3(bucket, 'thefile.zip'),
    });
    assertions_1.Template.fromStack(pipelineStack).hasResourceProperties('AWS::CodePipeline::Pipeline', {
        Stages: assertions_1.Match.arrayWith([{
                Name: 'Source',
                Actions: [
                    assertions_1.Match.objectLike({
                        Configuration: assertions_1.Match.objectLike({
                            S3Bucket: { Ref: assertions_1.Match.anyValue() },
                            S3ObjectKey: 'thefile.zip',
                        }),
                        Name: { Ref: assertions_1.Match.anyValue() },
                    }),
                ],
            }]),
    });
});
test('ECR source handles tokenized and namespaced names correctly', () => {
    const repository = new ecr.Repository(pipelineStack, 'Repository', { repositoryName: 'namespace/repo' });
    new testhelpers_1.ModernTestGitHubNpmPipeline(pipelineStack, 'Pipeline', {
        input: cdkp.CodePipelineSource.ecr(repository),
    });
    const template = assertions_1.Template.fromStack(pipelineStack);
    template.hasResourceProperties('AWS::CodePipeline::Pipeline', {
        Stages: assertions_1.Match.arrayWith([{
                Name: 'Source',
                Actions: [
                    assertions_1.Match.objectLike({
                        Configuration: assertions_1.Match.objectLike({
                            RepositoryName: { Ref: assertions_1.Match.anyValue() },
                        }),
                        Name: assertions_1.Match.objectLike({
                            'Fn::Join': [
                                '_',
                                {
                                    'Fn::Split': [
                                        '/',
                                        {
                                            Ref: assertions_1.Match.anyValue(),
                                        },
                                    ],
                                },
                            ],
                        }),
                    }),
                ],
            }]),
    });
});
test('GitHub source honors all valid properties', () => {
    new testhelpers_1.ModernTestGitHubNpmPipeline(pipelineStack, 'Pipeline', {
        input: cdkp.CodePipelineSource.gitHub('owner/repo', 'main', {
            trigger: aws_codepipeline_actions_1.GitHubTrigger.POLL,
            authentication: core_1.SecretValue.unsafePlainText('super-secret'),
        }),
    });
    assertions_1.Template.fromStack(pipelineStack).hasResourceProperties('AWS::CodePipeline::Pipeline', {
        Stages: assertions_1.Match.arrayWith([{
                Name: 'Source',
                Actions: [
                    assertions_1.Match.objectLike({
                        Configuration: assertions_1.Match.objectLike({
                            Owner: 'owner',
                            Repo: 'repo',
                            Branch: 'main',
                            PollForSourceChanges: true,
                            OAuthToken: 'super-secret',
                        }),
                        Name: 'owner_repo',
                    }),
                ],
            }]),
    });
});
test('GitHub source does not accept ill-formatted identifiers', () => {
    expect(() => {
        new testhelpers_1.ModernTestGitHubNpmPipeline(pipelineStack, 'Pipeline', {
            input: cdkp.CodePipelineSource.gitHub('repo-only', 'main'),
        });
    }).toThrow('GitHub repository name should be a resolved string like \'<owner>/<repo>\', got \'repo-only\'');
});
test('GitHub source does not accept unresolved identifiers', () => {
    expect(() => {
        new testhelpers_1.ModernTestGitHubNpmPipeline(pipelineStack, 'Pipeline', {
            input: cdkp.CodePipelineSource.gitHub(core_1.Token.asString({}), 'main'),
        });
    }).toThrow(/Step id cannot be unresolved/);
});
test('Dashes in repo names are removed from artifact names', () => {
    new testhelpers_1.ModernTestGitHubNpmPipeline(pipelineStack, 'Pipeline', {
        input: cdkp.CodePipelineSource.gitHub('owner/my-repo', 'main'),
    });
    assertions_1.Template.fromStack(pipelineStack).hasResourceProperties('AWS::CodePipeline::Pipeline', {
        Stages: assertions_1.Match.arrayWith([{
                Name: 'Source',
                Actions: [
                    assertions_1.Match.objectLike({
                        OutputArtifacts: [
                            { Name: 'owner_my_repo_Source' },
                        ],
                    }),
                ],
            }]),
    });
});
test('artifact names are never longer than 128 characters', () => {
    new testhelpers_1.ModernTestGitHubNpmPipeline(pipelineStack, 'Pipeline', {
        input: cdkp.CodePipelineSource.gitHub('owner/' + 'my-repo'.repeat(100), 'main'),
    });
    const artifactId = new assertions_1.Capture();
    assertions_1.Template.fromStack(pipelineStack).hasResourceProperties('AWS::CodePipeline::Pipeline', {
        Stages: assertions_1.Match.arrayWith([{
                Name: 'Source',
                Actions: [
                    assertions_1.Match.objectLike({
                        OutputArtifacts: [
                            { Name: artifactId },
                        ],
                    }),
                ],
            }]),
    });
    expect(artifactId.asString().length).toBeLessThanOrEqual(128);
});
test('can use source attributes in pipeline', () => {
    const gitHub = cdkp.CodePipelineSource.gitHub('owner/my-repo', 'main');
    // WHEN
    new testhelpers_1.ModernTestGitHubNpmPipeline(pipelineStack, 'Pipeline', {
        input: gitHub,
        synth: new cdkp.ShellStep('Synth', {
            env: {
                GITHUB_URL: gitHub.sourceAttribute('CommitUrl'),
            },
            commands: [
                'echo "Click here: $GITHUB_URL"',
            ],
        }),
        selfMutation: false,
    });
    assertions_1.Template.fromStack(pipelineStack).hasResourceProperties('AWS::CodePipeline::Pipeline', {
        Stages: [
            { Name: 'Source' },
            {
                Name: 'Build',
                Actions: [
                    {
                        Name: 'Synth',
                        Configuration: assertions_1.Match.objectLike({
                            EnvironmentVariables: assertions_1.Match.serializedJson([
                                {
                                    name: 'GITHUB_URL',
                                    type: 'PLAINTEXT',
                                    value: '#{Source@owner_my-repo.CommitUrl}',
                                },
                            ]),
                        }),
                    },
                ],
            },
        ],
    });
});
test('pass role to s3 codepipeline source', () => {
    const bucket = new s3.Bucket(pipelineStack, 'Bucket');
    const role = new aws_iam_1.Role(pipelineStack, 'TestRole', {
        assumedBy: new aws_iam_1.AnyPrincipal(),
    });
    new testhelpers_1.ModernTestGitHubNpmPipeline(pipelineStack, 'Pipeline', {
        input: cdkp.CodePipelineSource.s3(bucket, 'thefile.zip', {
            role,
        }),
    });
    assertions_1.Template.fromStack(pipelineStack).hasResourceProperties('AWS::CodePipeline::Pipeline', {
        Stages: assertions_1.Match.arrayWith([{
                Name: 'Source',
                Actions: [
                    assertions_1.Match.objectLike({
                        Configuration: assertions_1.Match.objectLike({
                            S3Bucket: { Ref: assertions_1.Match.anyValue() },
                            S3ObjectKey: 'thefile.zip',
                        }),
                        Name: { Ref: assertions_1.Match.anyValue() },
                        RoleArn: {
                            'Fn::GetAtt': [
                                assertions_1.Match.stringLikeRegexp('TestRole.*'),
                                'Arn',
                            ],
                        },
                    }),
                ],
            }]),
    });
});
test.each([
    ['CodeCommit', (stack) => {
            const repo = new ccommit.Repository(stack, 'Repo', {
                repositoryName: 'MyRepo',
            });
            return cdkp.CodePipelineSource.codeCommit(repo, 'main', {
                actionName: 'ConfiguredName',
            });
        }],
    ['S3', (stack) => {
            const bucket = new s3.Bucket(stack, 'Bucket');
            return cdkp.CodePipelineSource.s3(bucket, 'thefile.zip', {
                actionName: 'ConfiguredName',
            });
        }],
    ['ECR', (stack) => {
            const repository = new ecr.Repository(stack, 'Repository', { repositoryName: 'namespace/repo' });
            return cdkp.CodePipelineSource.ecr(repository, {
                actionName: 'ConfiguredName',
            });
        }],
    ['GitHub', () => {
            return cdkp.CodePipelineSource.gitHub('owner/repo', 'main', {
                actionName: 'ConfiguredName',
            });
        }],
    ['CodeStar', () => {
            return cdkp.CodePipelineSource.connection('owner/repo', 'main', {
                connectionArn: 'arn:aws:codestar-connections:us-west-2:123456789012:connection/39e4c34d-e13a-4e94-a886',
                actionName: 'ConfiguredName',
            });
        }],
])('can configure actionName for %s', (_name, fac) => {
    new testhelpers_1.ModernTestGitHubNpmPipeline(pipelineStack, 'Pipeline', {
        input: fac(pipelineStack),
    });
    assertions_1.Template.fromStack(pipelineStack).hasResourceProperties('AWS::CodePipeline::Pipeline', {
        Stages: assertions_1.Match.arrayWith([{
                Name: 'Source',
                Actions: [
                    assertions_1.Match.objectLike({
                        Name: 'ConfiguredName',
                    }),
                ],
            }]),
    });
});
