// aws-cdk-lib/aws-route53/test/delete-existing-record-set-handler.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const mockListResourceRecordSetsResponse = jest.fn();
const mockChangeResourceRecordSetsResponse = jest.fn();
const mockRoute53 = {
    listResourceRecordSets: jest.fn().mockReturnValue({
        promise: mockListResourceRecordSetsResponse,
    }),
    changeResourceRecordSets: jest.fn().mockReturnValue({
        promise: mockChangeResourceRecordSetsResponse,
    }),
    waitFor: jest.fn().mockReturnValue({
        promise: jest.fn().mockResolvedValue({}),
    }),
};
jest.mock('aws-sdk', () => {
    return {
        Route53: jest.fn(() => mockRoute53),
    };
});
const delete_existing_record_set_handler_1 = require("../lib/delete-existing-record-set-handler");
const event = {
    RequestType: 'Create',
    ServiceToken: 'service-token',
    ResponseURL: 'response-url',
    StackId: 'stack-id',
    RequestId: 'request-id',
    LogicalResourceId: 'logical-resource-id',
    ResourceType: 'Custom::DeleteExistingRecordSet',
    ResourceProperties: {
        ServiceToken: 'service-token',
        HostedZoneId: 'hosted-zone-id',
        RecordName: 'dev.cdk.aws.',
        RecordType: 'A',
    },
};
beforeEach(() => {
    jest.clearAllMocks();
});
test('create request with existing record', async () => {
    mockListResourceRecordSetsResponse.mockResolvedValueOnce({
        ResourceRecordSets: [
            {
                Name: 'dev.cdk.aws.',
                Type: 'A',
                TTL: 900,
            },
            {
                Name: 'dev.cdk.aws.',
                Type: 'AAAA',
                TTL: 900,
            },
        ],
    });
    mockChangeResourceRecordSetsResponse.mockResolvedValueOnce({
        ChangeInfo: {
            Id: 'change-id',
        },
    });
    await (0, delete_existing_record_set_handler_1.handler)(event);
    expect(mockRoute53.listResourceRecordSets).toHaveBeenCalledWith({
        HostedZoneId: 'hosted-zone-id',
        StartRecordName: 'dev.cdk.aws.',
        StartRecordType: 'A',
    });
    expect(mockRoute53.changeResourceRecordSets).toHaveBeenCalledWith({
        HostedZoneId: 'hosted-zone-id',
        ChangeBatch: {
            Changes: [
                {
                    Action: 'DELETE',
                    ResourceRecordSet: {
                        Name: 'dev.cdk.aws.',
                        TTL: 900,
                        Type: 'A',
                    },
                },
            ],
        },
    });
    expect(mockRoute53.waitFor).toHaveBeenCalledWith('resourceRecordSetsChanged', {
        Id: 'change-id',
    });
});
test('create request with non existing record', async () => {
    mockListResourceRecordSetsResponse.mockResolvedValueOnce({
        ResourceRecordSets: [
            {
                Name: 'www.cdk.aws.',
                Type: 'A',
                TTL: 900,
            },
            {
                Name: 'dev.cdk.aws.',
                Type: 'MX',
                TTL: 900,
            },
        ],
    });
    await (0, delete_existing_record_set_handler_1.handler)(event);
    expect(mockRoute53.changeResourceRecordSets).not.toHaveBeenCalled();
});
test('update request', async () => {
    await (0, delete_existing_record_set_handler_1.handler)({
        ...event,
        RequestType: 'Update',
        PhysicalResourceId: 'id',
        OldResourceProperties: {},
    });
    expect(mockRoute53.changeResourceRecordSets).not.toHaveBeenCalled();
});
test('delete request', async () => {
    await (0, delete_existing_record_set_handler_1.handler)({
        ...event,
        RequestType: 'Delete',
        PhysicalResourceId: 'id',
    });
    expect(mockRoute53.changeResourceRecordSets).not.toHaveBeenCalled();
});
test('with alias target', async () => {
    mockListResourceRecordSetsResponse.mockResolvedValueOnce({
        ResourceRecordSets: [
            {
                Name: 'dev.cdk.aws.',
                Type: 'A',
                TTL: undefined,
                ResourceRecords: [],
                AliasTarget: {
                    HostedZoneId: 'hosted-zone-id',
                    DNSName: 'dns-name',
                    EvaluateTargetHealth: false,
                },
            },
        ],
    });
    mockChangeResourceRecordSetsResponse.mockResolvedValueOnce({
        ChangeInfo: {
            Id: 'change-id',
        },
    });
    await (0, delete_existing_record_set_handler_1.handler)(event);
    expect(mockRoute53.changeResourceRecordSets).toHaveBeenCalledWith({
        HostedZoneId: 'hosted-zone-id',
        ChangeBatch: {
            Changes: [
                {
                    Action: 'DELETE',
                    ResourceRecordSet: {
                        Name: 'dev.cdk.aws.',
                        Type: 'A',
                        AliasTarget: {
                            HostedZoneId: 'hosted-zone-id',
                            DNSName: 'dns-name',
                            EvaluateTargetHealth: false,
                        },
                    },
                },
            ],
        },
    });
});
