// aws-cdk-lib/aws-stepfunctions/test/state.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const assert = require("../../assertions");
const cdk = require("../../core");
const fake_task_1 = require("./fake-task");
const lib_1 = require("../lib");
test('JsonPath.DISCARD can be used to discard a state\'s output', () => {
    // GIVEN
    const app = new cdk.App();
    const stack = new cdk.Stack(app, 'TestStack');
    const task = new fake_task_1.FakeTask(stack, 'my-state', {
        inputPath: lib_1.JsonPath.DISCARD,
        outputPath: lib_1.JsonPath.DISCARD,
        resultPath: lib_1.JsonPath.DISCARD,
    });
    new lib_1.StateMachine(stack, 'state-machine', {
        definition: task,
    });
    // WHEN
    const definitionString = new assert.Capture();
    assert.Template.fromStack(stack).hasResourceProperties('AWS::StepFunctions::StateMachine', {
        DefinitionString: definitionString,
    });
    // THEN
    const definition = JSON.parse(definitionString.asString());
    expect(definition).toMatchObject({
        States: {
            'my-state': {
                InputPath: null,
                OutputPath: null,
                ResultPath: null,
            },
        },
    });
});
