// aws-cdk-lib/aws-stepfunctions/test/wait.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const cdk = require("../../core");
const render_util_1 = require("./private/render-util");
const lib_1 = require("../lib");
describe('Wait State', () => {
    test('wait time from ISO8601 timestamp', () => {
        // GIVEN
        const timestamp = '2025-01-01T00:00:00Z';
        // WHEN
        const waitTime = lib_1.WaitTime.timestamp(timestamp);
        // THEN
        expect(waitTime).toEqual({
            json: {
                Timestamp: '2025-01-01T00:00:00Z',
            },
        });
    });
    test('wait time from seconds path in state object', () => {
        // GIVEN
        const secondsPath = '$.waitSeconds';
        // WHEN
        const waitTime = lib_1.WaitTime.secondsPath(secondsPath);
        // THEN
        expect(waitTime).toEqual({
            json: {
                SecondsPath: '$.waitSeconds',
            },
        });
    });
    test('wait time from timestamp path in state object', () => {
        // GIVEN
        const path = '$.timestampPath';
        // WHEN
        const waitTime = lib_1.WaitTime.timestampPath(path);
        // THEN
        expect(waitTime).toEqual({
            json: {
                TimestampPath: '$.timestampPath',
            },
        });
    });
    test('supports adding a next state', () => {
        // GIVEN
        const stack = new cdk.Stack();
        const chain = new lib_1.Wait(stack, 'myWaitState', {
            time: lib_1.WaitTime.duration(cdk.Duration.seconds(30)),
        });
        // WHEN
        chain.next(new lib_1.Pass(stack, 'final pass', {}));
        // THEN
        expect((0, render_util_1.render)(stack, chain)).toEqual({
            StartAt: 'myWaitState',
            States: {
                'final pass': {
                    End: true,
                    Type: 'Pass',
                },
                'myWaitState': {
                    Next: 'final pass',
                    Seconds: 30,
                    Type: 'Wait',
                },
            },
        });
    });
});
