// aws-cdk-lib/aws-elasticsearch/test/log-group-resource-policy.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const assertions_1 = require("../../assertions");
const iam = require("../../aws-iam");
const core_1 = require("../../core");
const log_group_resource_policy_1 = require("../lib/log-group-resource-policy");
let app;
let stack;
beforeEach(() => {
    app = new core_1.App();
    stack = new core_1.Stack(app, 'Stack', {
        env: { account: '1234', region: 'testregion' },
    });
});
test('minimal example renders correctly', () => {
    new log_group_resource_policy_1.LogGroupResourcePolicy(stack, 'LogGroupResourcePolicy', {
        policyName: 'TestPolicy',
        policyStatements: [new iam.PolicyStatement({
                effect: iam.Effect.ALLOW,
                actions: ['logs:PutLogEvents', 'logs:CreateLogStream'],
                resources: ['*'],
                principals: [new iam.ServicePrincipal('es.amazonaws.com')],
            })],
    });
    assertions_1.Template.fromStack(stack).hasResourceProperties('Custom::CloudwatchLogResourcePolicy', {
        ServiceToken: {
            'Fn::GetAtt': [
                'AWS679f53fac002430cb0da5b7982bd22872D164C4C',
                'Arn',
            ],
        },
        Create: JSON.stringify({
            service: 'CloudWatchLogs',
            action: 'putResourcePolicy',
            parameters: {
                policyName: 'TestPolicy',
                policyDocument: '{"Statement":[{"Action":["logs:PutLogEvents","logs:CreateLogStream"],"Effect":"Allow","Principal":{"Service":"es.amazonaws.com"},"Resource":"*"}],"Version":"2012-10-17"}',
            },
            physicalResourceId: {
                id: 'LogGroupResourcePolicy',
            },
        }),
        Update: JSON.stringify({
            service: 'CloudWatchLogs',
            action: 'putResourcePolicy',
            parameters: {
                policyName: 'TestPolicy',
                policyDocument: '{"Statement":[{"Action":["logs:PutLogEvents","logs:CreateLogStream"],"Effect":"Allow","Principal":{"Service":"es.amazonaws.com"},"Resource":"*"}],"Version":"2012-10-17"}',
            },
            physicalResourceId: {
                id: 'LogGroupResourcePolicy',
            },
        }),
        Delete: JSON.stringify({
            service: 'CloudWatchLogs',
            action: 'deleteResourcePolicy',
            parameters: {
                policyName: 'TestPolicy',
            },
            ignoreErrorCodesMatching: 'ResourceNotFoundException',
        }),
    });
});
