// aws-cdk-lib/aws-ses/test/configuration-set.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const assertions_1 = require("../../assertions");
const core_1 = require("../../core");
const lib_1 = require("../lib");
let stack;
beforeEach(() => {
    stack = new core_1.Stack();
});
test('default configuration set', () => {
    new lib_1.ConfigurationSet(stack, 'ConfigurationSet');
    assertions_1.Template.fromStack(stack).hasResourceProperties('AWS::SES::ConfigurationSet', {
        Name: assertions_1.Match.absent(),
    });
});
test('configuration set with options', () => {
    new lib_1.ConfigurationSet(stack, 'ConfigurationSet', {
        customTrackingRedirectDomain: 'track.cdk.dev',
        suppressionReasons: lib_1.SuppressionReasons.COMPLAINTS_ONLY,
        tlsPolicy: lib_1.ConfigurationSetTlsPolicy.REQUIRE,
        dedicatedIpPool: new lib_1.DedicatedIpPool(stack, 'Pool'),
    });
    assertions_1.Template.fromStack(stack).hasResourceProperties('AWS::SES::ConfigurationSet', {
        DeliveryOptions: {
            SendingPoolName: {
                Ref: 'PoolD3F588B8',
            },
            TlsPolicy: 'REQUIRE',
        },
        SuppressionOptions: {
            SuppressedReasons: [
                'COMPLAINT',
            ],
        },
        TrackingOptions: {
            CustomRedirectDomain: 'track.cdk.dev',
        },
    });
});
