// aws-cdk-lib/aws-ses/test/receipt-rule-set.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const assertions_1 = require("../../assertions");
const core_1 = require("../../core");
const lib_1 = require("../lib");
/* eslint-disable quote-props */
describe('receipt rule set', () => {
    test('can create a receipt rule set', () => {
        // GIVEN
        const stack = new core_1.Stack();
        // WHEN
        new lib_1.ReceiptRuleSet(stack, 'RuleSet', {
            receiptRuleSetName: 'MyRuleSet',
        });
        // THEN
        assertions_1.Template.fromStack(stack).hasResourceProperties('AWS::SES::ReceiptRuleSet', {
            RuleSetName: 'MyRuleSet',
        });
    });
    test('can create a receipt rule set with drop spam', () => {
        // GIVEN
        const stack = new core_1.Stack();
        // WHEN
        new lib_1.ReceiptRuleSet(stack, 'RuleSet', {
            dropSpam: true,
        });
        // THEN
        assertions_1.Template.fromStack(stack).hasResourceProperties('AWS::SES::ReceiptRule', {
            Rule: {
                Actions: [
                    {
                        LambdaAction: {
                            FunctionArn: {
                                'Fn::GetAtt': [
                                    'SingletonLambda224e77f9a32e4b4dac32983477abba164533EA15',
                                    'Arn',
                                ],
                            },
                            InvocationType: 'RequestResponse',
                        },
                    },
                ],
                Enabled: true,
                ScanEnabled: true,
            },
        });
        assertions_1.Template.fromStack(stack).resourceCountIs('AWS::Lambda::Function', 1);
    });
    test('drop spam rule should always appear first', () => {
        // GIVEN
        const stack = new core_1.Stack();
        // WHEN
        new lib_1.ReceiptRuleSet(stack, 'RuleSet', {
            dropSpam: true,
            rules: [
                {
                    scanEnabled: true,
                    recipients: ['foo@example.com'],
                },
            ],
        });
        // THEN
        assertions_1.Template.fromStack(stack).hasResourceProperties('AWS::SES::ReceiptRule', {
            Rule: {
                Enabled: true,
                Recipients: [
                    'foo@example.com',
                ],
                ScanEnabled: true,
            },
            // All "regular" rules should come after the drop spam rule
            After: {
                Ref: 'RuleSetDropSpamRule5809F51B',
            },
        });
        assertions_1.Template.fromStack(stack).resourceCountIs('AWS::Lambda::Function', 1);
    });
    test('import receipt rule set', () => {
        // GIVEN
        const stack = new core_1.Stack();
        // WHEN
        const receiptRuleSet = lib_1.ReceiptRuleSet.fromReceiptRuleSetName(stack, 'ImportedRuleSet', 'MyRuleSet');
        receiptRuleSet.addRule('MyRule');
        // THEN
        assertions_1.Template.fromStack(stack).templateMatches({
            'Resources': {
                'ImportedRuleSetMyRule53EE2F7F': {
                    'Type': 'AWS::SES::ReceiptRule',
                    'Properties': {
                        'Rule': {
                            'Enabled': true,
                        },
                        'RuleSetName': 'MyRuleSet',
                    },
                },
            },
        });
    });
});
