// aws-cdk-lib/aws-stepfunctions-tasks/test/lambda/invoke-function.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const assertions_1 = require("../../../assertions");
const lambda = require("../../../aws-lambda");
const sfn = require("../../../aws-stepfunctions");
const cdk_build_tools_1 = require("@aws-cdk/cdk-build-tools");
const core_1 = require("../../../core");
const tasks = require("../../lib");
let stack;
let fn;
beforeEach(() => {
    stack = new core_1.Stack();
    fn = new lambda.Function(stack, 'Fn', {
        code: lambda.Code.fromInline('hello'),
        handler: 'index.hello',
        runtime: lambda.Runtime.PYTHON_3_9,
    });
});
(0, cdk_build_tools_1.describeDeprecated)('InvokeFunction', () => {
    test('Invoke lambda with function ARN', () => {
        // WHEN
        const task = new sfn.Task(stack, 'Task', { task: new tasks.InvokeFunction(fn) });
        new sfn.StateMachine(stack, 'SM', {
            definition: task,
        });
        // THEN
        assertions_1.Template.fromStack(stack).hasResourceProperties('AWS::StepFunctions::StateMachine', {
            DefinitionString: {
                'Fn::Join': ['', [
                        '{"StartAt":"Task","States":{"Task":{"End":true,"Type":"Task","Resource":"',
                        { 'Fn::GetAtt': ['Fn9270CBC0', 'Arn'] },
                        '"}}}',
                    ]],
            },
        });
    });
    test('Lambda function payload ends up in Parameters', () => {
        new sfn.StateMachine(stack, 'SM', {
            definition: new sfn.Task(stack, 'Task', {
                task: new tasks.InvokeFunction(fn, {
                    payload: {
                        foo: sfn.JsonPath.stringAt('$.bar'),
                    },
                }),
            }),
        });
        assertions_1.Template.fromStack(stack).hasResourceProperties('AWS::StepFunctions::StateMachine', {
            DefinitionString: {
                'Fn::Join': ['', [
                        '{"StartAt":"Task","States":{"Task":{"End":true,"Parameters":{"foo.$":"$.bar"},"Type":"Task","Resource":"',
                        { 'Fn::GetAtt': ['Fn9270CBC0', 'Arn'] },
                        '"}}}',
                    ]],
            },
        });
    });
});
