// aws-cdk-lib/aws-codepipeline-actions/test/s3/s3-deploy-action.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const assertions_1 = require("../../../assertions");
const codepipeline = require("../../../aws-codepipeline");
const s3 = require("../../../aws-s3");
const core_1 = require("../../../core");
const cpactions = require("../../lib");
/* eslint-disable quote-props */
describe('S3 Deploy Action', () => {
    test('by default extract artifacts', () => {
        const stack = new core_1.Stack();
        minimalPipeline(stack);
        assertions_1.Template.fromStack(stack).hasResourceProperties('AWS::CodePipeline::Pipeline', {
            'Stages': [
                {
                    'Name': 'Source',
                    'Actions': [
                        {
                            'Name': 'Source',
                            'ActionTypeId': {
                                'Category': 'Source',
                                'Owner': 'ThirdParty',
                            },
                        },
                    ],
                },
                {
                    'Name': 'Deploy',
                    'Actions': [
                        {
                            'ActionTypeId': {
                                'Category': 'Deploy',
                                'Provider': 'S3',
                            },
                            'Configuration': {
                                'Extract': 'true',
                            },
                            'Name': 'CopyFiles',
                        },
                    ],
                },
            ],
        });
    });
    test('grant the pipeline correct access to the target bucket', () => {
        const stack = new core_1.Stack();
        minimalPipeline(stack);
        assertions_1.Template.fromStack(stack).hasResourceProperties('AWS::IAM::Policy', {
            'PolicyDocument': {
                'Statement': [
                    {
                        'Effect': 'Allow',
                        'Action': [
                            's3:GetObject*',
                            's3:GetBucket*',
                            's3:List*',
                            's3:DeleteObject*',
                            's3:PutObject',
                            's3:PutObjectLegalHold',
                            's3:PutObjectRetention',
                            's3:PutObjectTagging',
                            's3:PutObjectVersionTagging',
                            's3:Abort*',
                        ],
                    },
                    {},
                    {
                        'Effect': 'Allow',
                        'Action': 'sts:AssumeRole',
                    },
                ],
            },
        });
    });
    test('kebab-case CannedACL value', () => {
        const stack = new core_1.Stack();
        minimalPipeline(stack, {
            accessControl: s3.BucketAccessControl.PUBLIC_READ_WRITE,
        });
        assertions_1.Template.fromStack(stack).hasResourceProperties('AWS::CodePipeline::Pipeline', {
            'Stages': [
                {},
                {
                    'Actions': [
                        {
                            'Configuration': {
                                'CannedACL': 'public-read-write',
                            },
                        },
                    ],
                },
            ],
        });
    });
    test('allow customizing cache-control', () => {
        const stack = new core_1.Stack();
        minimalPipeline(stack, {
            cacheControl: [
                cpactions.CacheControl.setPublic(),
                cpactions.CacheControl.maxAge(core_1.Duration.hours(12)),
                cpactions.CacheControl.sMaxAge(core_1.Duration.hours(12)),
            ],
        });
        assertions_1.Template.fromStack(stack).hasResourceProperties('AWS::CodePipeline::Pipeline', {
            'Stages': [
                {},
                {
                    'Actions': [
                        {
                            'Configuration': {
                                'CacheControl': 'public, max-age=43200, s-maxage=43200',
                            },
                        },
                    ],
                },
            ],
        });
    });
    test('allow customizing objectKey (deployment path on S3)', () => {
        const stack = new core_1.Stack();
        minimalPipeline(stack, {
            objectKey: '/a/b/c',
        });
        assertions_1.Template.fromStack(stack).hasResourceProperties('AWS::CodePipeline::Pipeline', {
            'Stages': [
                {},
                {
                    'Actions': [
                        {
                            'Configuration': {
                                'ObjectKey': '/a/b/c',
                            },
                        },
                    ],
                },
            ],
        });
    });
    test('correctly makes the action cross-region for a Bucket imported with a different region', () => {
        const app = new core_1.App();
        const stack = new core_1.Stack(app, 'PipelineStack', {
            env: { account: '123456789012', region: 'us-west-2' },
        });
        const deployBucket = s3.Bucket.fromBucketAttributes(stack, 'DeployBucket', {
            bucketName: 'my-deploy-bucket',
            region: 'ap-southeast-1',
        });
        minimalPipeline(stack, {
            bucket: deployBucket,
        });
        assertions_1.Template.fromStack(stack).hasResourceProperties('AWS::CodePipeline::Pipeline', {
            Stages: [
                {},
                {
                    Name: 'Deploy',
                    Actions: [
                        {
                            Name: 'CopyFiles',
                            Region: 'ap-southeast-1',
                        },
                    ],
                },
            ],
        });
    });
});
function minimalPipeline(stack, options = {}) {
    const sourceOutput = new codepipeline.Artifact();
    const sourceAction = new cpactions.GitHubSourceAction({
        actionName: 'Source',
        owner: 'aws',
        repo: 'aws-cdk',
        output: sourceOutput,
        oauthToken: core_1.SecretValue.unsafePlainText('secret'),
    });
    const pipeline = new codepipeline.Pipeline(stack, 'MyPipeline', {
        stages: [
            {
                stageName: 'Source',
                actions: [sourceAction],
            },
        ],
    });
    const deploymentStage = pipeline.addStage({
        stageName: 'Deploy',
        actions: [
            new cpactions.S3DeployAction({
                accessControl: options.accessControl,
                actionName: 'CopyFiles',
                bucket: options.bucket || new s3.Bucket(stack, 'MyBucket'),
                cacheControl: options.cacheControl,
                extract: options.extract,
                input: sourceOutput,
                objectKey: options.objectKey,
            }),
        ],
    });
    return deploymentStage;
}
