// aws-cdk-lib/aws-ecs/test/ec2/cross-stack.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const assertions_1 = require("../../../assertions");
const ec2 = require("../../../aws-ec2");
const elbv2 = require("../../../aws-elasticloadbalancingv2");
const core_1 = require("../../../core");
const ecs = require("../../lib");
const util_1 = require("../util");
// Test various cross-stack Cluster/Service/ALB scenario's
let app;
let stack1;
let stack2;
let cluster;
let service;
describe('cross stack', () => {
    beforeEach(() => {
        app = new core_1.App();
        stack1 = new core_1.Stack(app, 'Stack1');
        const vpc = new ec2.Vpc(stack1, 'Vpc');
        cluster = new ecs.Cluster(stack1, 'Cluster', {
            vpc,
        });
        (0, util_1.addDefaultCapacityProvider)(cluster, stack1, vpc);
        stack2 = new core_1.Stack(app, 'Stack2');
        const taskDefinition = new ecs.Ec2TaskDefinition(stack2, 'TD');
        const container = taskDefinition.addContainer('Main', {
            image: ecs.ContainerImage.fromRegistry('asdf'),
            memoryLimitMiB: 512,
        });
        container.addPortMappings({ containerPort: 8000 });
        service = new ecs.Ec2Service(stack2, 'Service', {
            cluster,
            taskDefinition,
        });
    });
    test('ALB next to Service', () => {
        // WHEN
        const lb = new elbv2.ApplicationLoadBalancer(stack2, 'ALB', { vpc: cluster.vpc });
        const listener = lb.addListener('listener', { port: 80 });
        listener.addTargets('target', {
            port: 80,
            targets: [service],
        });
        // THEN: it shouldn't throw due to cyclic dependencies
        assertions_1.Template.fromStack(stack2).resourceCountIs('AWS::ECS::Service', 1);
        expectIngress(stack2);
    });
    test('ALB next to Cluster', () => {
        // WHEN
        const lb = new elbv2.ApplicationLoadBalancer(stack1, 'ALB', { vpc: cluster.vpc });
        const listener = lb.addListener('listener', { port: 80 });
        listener.addTargets('target', {
            port: 80,
            targets: [service],
        });
        // THEN: it shouldn't throw due to cyclic dependencies
        assertions_1.Template.fromStack(stack2).resourceCountIs('AWS::ECS::Service', 1);
        expectIngress(stack2);
    });
    test('ALB in its own stack', () => {
        // WHEN
        const stack3 = new core_1.Stack(app, 'Stack3');
        const lb = new elbv2.ApplicationLoadBalancer(stack3, 'ALB', { vpc: cluster.vpc });
        const listener = lb.addListener('listener', { port: 80 });
        listener.addTargets('target', {
            port: 80,
            targets: [service],
        });
        // THEN: it shouldn't throw due to cyclic dependencies
        assertions_1.Template.fromStack(stack2).resourceCountIs('AWS::ECS::Service', 1);
        expectIngress(stack2);
    });
});
function expectIngress(stack) {
    assertions_1.Template.fromStack(stack).hasResourceProperties('AWS::EC2::SecurityGroupIngress', {
        FromPort: 32768,
        ToPort: 65535,
        GroupId: { 'Fn::ImportValue': 'Stack1:ExportsOutputFnGetAttDefaultAutoScalingGroupInstanceSecurityGroupFBA881D0GroupId2F7C804A' },
    });
}
