// aws-cdk-lib/aws-stepfunctions-tasks/test/databrew/start-job-run.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const sfn = require("../../../aws-stepfunctions");
const cdk = require("../../../core");
const start_job_run_1 = require("../../lib/databrew/start-job-run");
describe('Start Job Run', () => {
    test('default settings', () => {
        // GIVEN
        const stack = new cdk.Stack();
        // WHEN
        const task = new start_job_run_1.GlueDataBrewStartJobRun(stack, 'JobRun', {
            name: 'jobName',
        });
        // THEN
        expect(stack.resolve(task.toStateJson())).toEqual({
            Type: 'Task',
            Resource: {
                'Fn::Join': [
                    '',
                    [
                        'arn:',
                        {
                            Ref: 'AWS::Partition',
                        },
                        ':states:::databrew:startJobRun',
                    ],
                ],
            },
            End: true,
            Parameters: {
                Name: 'jobName',
            },
        });
    });
    test('create job with input from task', () => {
        // GIVEN
        const stack = new cdk.Stack();
        // WHEN
        const task = new start_job_run_1.GlueDataBrewStartJobRun(stack, 'JobRun', {
            name: sfn.JsonPath.stringAt('$.Name'),
        });
        // THEN
        expect(stack.resolve(task.toStateJson())).toEqual({
            Type: 'Task',
            Resource: {
                'Fn::Join': [
                    '',
                    [
                        'arn:',
                        {
                            Ref: 'AWS::Partition',
                        },
                        ':states:::databrew:startJobRun',
                    ],
                ],
            },
            End: true,
            Parameters: {
                'Name.$': '$.Name',
            },
        });
    });
    test('sync integrationPattern', () => {
        // GIVEN
        const stack = new cdk.Stack();
        // WHEN
        const task = new start_job_run_1.GlueDataBrewStartJobRun(stack, 'JobRun', {
            integrationPattern: sfn.IntegrationPattern.RUN_JOB,
            name: 'jobName',
        });
        // THEN
        expect(stack.resolve(task.toStateJson())).toEqual({
            Type: 'Task',
            Resource: {
                'Fn::Join': [
                    '',
                    [
                        'arn:',
                        {
                            Ref: 'AWS::Partition',
                        },
                        ':states:::databrew:startJobRun.sync',
                    ],
                ],
            },
            End: true,
            Parameters: {
                Name: 'jobName',
            },
        });
    });
    test('wait_for_task_token integrationPattern throws an error', () => {
        // GIVEN
        const stack = new cdk.Stack();
        expect(() => {
            new start_job_run_1.GlueDataBrewStartJobRun(stack, 'JobRun', {
                integrationPattern: sfn.IntegrationPattern.WAIT_FOR_TASK_TOKEN,
                name: 'jobName',
            });
        }).toThrow(/Unsupported service integration pattern. Supported Patterns: REQUEST_RESPONSE,RUN_JOB. Received: WAIT_FOR_TASK_TOKEN/i);
    });
});
