// aws-cdk-lib/aws-appsync/test/appsync-mapping-template.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const path = require("path");
const assertions_1 = require("../../assertions");
const lambda = require("../../aws-lambda");
const cdk = require("../../core");
const appsync = require("../lib");
let stack;
let api;
beforeEach(() => {
    // GIVEN
    stack = new cdk.Stack();
    api = new appsync.GraphqlApi(stack, 'api', {
        name: 'api',
        schema: appsync.SchemaFile.fromAsset(path.join(__dirname, 'appsync.lambda.graphql')),
    });
});
describe('Lambda Mapping Templates', () => {
    // GIVEN
    let func;
    const invokeMT = '{"version": "2017-02-28", "operation": "Invoke", "payload": $util.toJson($ctx)}';
    const batchMT = '{"version": "2017-02-28", "operation": "BatchInvoke", "payload": $util.toJson($ctx)}';
    beforeEach(() => {
        func = new lambda.Function(stack, 'func', {
            code: lambda.Code.fromAsset(path.join(__dirname, 'verify/lambda-tutorial')),
            handler: 'lambda-tutorial.handler',
            runtime: lambda.Runtime.NODEJS_14_X,
        });
    });
    test('Lambda request default operation is "Invoke"', () => {
        // WHEN
        const lambdaDS = api.addLambdaDataSource('LambdaDS', func);
        lambdaDS.createResolver('QueryAllPosts', {
            typeName: 'Query',
            fieldName: 'allPosts',
            requestMappingTemplate: appsync.MappingTemplate.lambdaRequest(),
            responseMappingTemplate: appsync.MappingTemplate.lambdaResult(),
        });
        // THEN
        assertions_1.Template.fromStack(stack).hasResourceProperties('AWS::AppSync::Resolver', {
            FieldName: 'allPosts',
            RequestMappingTemplate: invokeMT,
        });
    });
    test('Lambda request supports "BatchInvoke" through custom operation', () => {
        // WHEN
        const lambdaDS = api.addLambdaDataSource('LambdaDS', func);
        lambdaDS.createResolver('PostRelatedPosts', {
            typeName: 'Post',
            fieldName: 'relatedPosts',
            requestMappingTemplate: appsync.MappingTemplate.lambdaRequest('$util.toJson($ctx)', 'BatchInvoke'),
            responseMappingTemplate: appsync.MappingTemplate.lambdaResult(),
            maxBatchSize: 10,
        });
        // THEN
        assertions_1.Template.fromStack(stack).hasResourceProperties('AWS::AppSync::Resolver', {
            FieldName: 'relatedPosts',
            RequestMappingTemplate: batchMT,
            MaxBatchSize: 10,
        });
    });
});
