// aws-cdk-lib/aws-cloudfront/test/oai.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const assertions_1 = require("../../assertions");
const cdk_build_tools_1 = require("@aws-cdk/cdk-build-tools");
const cdk = require("../../core");
const lib_1 = require("../lib");
describe('Origin Access Identity', () => {
    test('With automatic comment', () => {
        const stack = new cdk.Stack();
        new lib_1.OriginAccessIdentity(stack, 'OAI');
        assertions_1.Template.fromStack(stack).templateMatches({
            Resources: {
                OAIE1EFC67F: {
                    Type: 'AWS::CloudFront::CloudFrontOriginAccessIdentity',
                    Properties: {
                        CloudFrontOriginAccessIdentityConfig: {
                            Comment: 'Allows CloudFront to reach the bucket',
                        },
                    },
                },
            },
        });
    });
    test('With provided comment', () => {
        const stack = new cdk.Stack();
        new lib_1.OriginAccessIdentity(stack, 'OAI', {
            comment: 'test comment',
        });
        assertions_1.Template.fromStack(stack).templateMatches({
            Resources: {
                OAIE1EFC67F: {
                    Type: 'AWS::CloudFront::CloudFrontOriginAccessIdentity',
                    Properties: {
                        CloudFrontOriginAccessIdentityConfig: {
                            Comment: 'test comment',
                        },
                    },
                },
            },
        });
    });
    test('Truncates long comments', () => {
        const stack = new cdk.Stack();
        new lib_1.OriginAccessIdentity(stack, 'OAI', {
            comment: 'This is a really long comment. Auto-generated comments based on ids of origins might sometimes be this long or even longer and that will break',
        });
        assertions_1.Template.fromStack(stack).hasResourceProperties('AWS::CloudFront::CloudFrontOriginAccessIdentity', {
            CloudFrontOriginAccessIdentityConfig: {
                Comment: 'This is a really long comment. Auto-generated comments based on ids of origins might sometimes be this long or even longer and t',
            },
        });
    });
    (0, cdk_build_tools_1.testDeprecated)('Builds ARN of CloudFront user for fromOriginAccessIdentityName', () => {
        const stack = new cdk.Stack();
        const oai = lib_1.OriginAccessIdentity.fromOriginAccessIdentityName(stack, 'OAI', 'OAITest');
        expect(oai.grantPrincipal.policyFragment.principalJson.AWS[0]).toMatch(/:iam::cloudfront:user\/CloudFront Origin Access Identity OAITest$/);
    });
    test('Builds ARN of CloudFront user for fromOriginAccessIdentityId', () => {
        const stack = new cdk.Stack();
        const oai = lib_1.OriginAccessIdentity.fromOriginAccessIdentityId(stack, 'OAI', 'OAITest');
        expect(oai.grantPrincipal.policyFragment.principalJson.AWS[0]).toMatch(/:iam::cloudfront:user\/CloudFront Origin Access Identity OAITest$/);
    });
});
