// aws-cdk-lib/aws-events-targets/test/logs/log-group.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const assertions_1 = require("../../../assertions");
const events = require("../../../aws-events");
const logs = require("../../../aws-logs");
const sqs = require("../../../aws-sqs");
const cdk_build_tools_1 = require("@aws-cdk/cdk-build-tools");
const cdk = require("../../../core");
const targets = require("../../lib");
const lib_1 = require("../../lib");
test('use log group as an event rule target', () => {
    // GIVEN
    const stack = new cdk.Stack();
    const logGroup = new logs.LogGroup(stack, 'MyLogGroup', {
        logGroupName: '/aws/events/MyLogGroup',
    });
    const rule1 = new events.Rule(stack, 'Rule', {
        schedule: events.Schedule.rate(cdk.Duration.minutes(1)),
    });
    // WHEN
    rule1.addTarget(new targets.CloudWatchLogGroup(logGroup));
    // THEN
    assertions_1.Template.fromStack(stack).hasResourceProperties('AWS::Events::Rule', {
        ScheduleExpression: 'rate(1 minute)',
        State: 'ENABLED',
        Targets: [
            {
                Arn: {
                    'Fn::Join': [
                        '',
                        [
                            'arn:',
                            {
                                Ref: 'AWS::Partition',
                            },
                            ':logs:',
                            {
                                Ref: 'AWS::Region',
                            },
                            ':',
                            {
                                Ref: 'AWS::AccountId',
                            },
                            ':log-group:',
                            {
                                Ref: 'MyLogGroup5C0DAD85',
                            },
                        ],
                    ],
                },
                Id: 'Target0',
            },
        ],
    });
});
(0, cdk_build_tools_1.testDeprecated)('use log group as an event rule target with rule target input', () => {
    // GIVEN
    const app = new cdk.App();
    const stack = new cdk.Stack(app);
    const logGroup = new logs.LogGroup(stack, 'MyLogGroup', {
        logGroupName: '/aws/events/MyLogGroup',
    });
    const rule1 = new events.Rule(stack, 'Rule', {
        schedule: events.Schedule.rate(cdk.Duration.minutes(1)),
    });
    // WHEN
    rule1.addTarget(new targets.CloudWatchLogGroup(logGroup, {
        event: events.RuleTargetInput.fromObject({
            message: events.EventField.fromPath('$'),
        }),
    }));
    // THEN
    expect(() => {
        app.synth();
    }).toThrow(/CloudWatchLogGroup targets only support input templates in the format/);
});
(0, cdk_build_tools_1.testDeprecated)('cannot use both logEvent and event', () => {
    // GIVEN
    const app = new cdk.App();
    const stack = new cdk.Stack(app);
    const logGroup = new logs.LogGroup(stack, 'MyLogGroup', {
        logGroupName: '/aws/events/MyLogGroup',
    });
    const rule1 = new events.Rule(stack, 'Rule', {
        schedule: events.Schedule.rate(cdk.Duration.minutes(1)),
    });
    // THEN
    expect(() => {
        rule1.addTarget(new targets.CloudWatchLogGroup(logGroup, {
            event: events.RuleTargetInput.fromObject({
                message: events.EventField.fromPath('$'),
            }),
            logEvent: lib_1.LogGroupTargetInput.fromObject(),
        }));
    }).toThrow(/Only one of "event" or "logEvent" can be specified/);
});
test('logEvent with defaults', () => {
    // GIVEN
    const stack = new cdk.Stack();
    const logGroup = new logs.LogGroup(stack, 'MyLogGroup', {
        logGroupName: '/aws/events/MyLogGroup',
    });
    const rule1 = new events.Rule(stack, 'Rule', {
        schedule: events.Schedule.rate(cdk.Duration.minutes(1)),
    });
    // WHEN
    rule1.addTarget(new targets.CloudWatchLogGroup(logGroup, {
        logEvent: lib_1.LogGroupTargetInput.fromObject(),
    }));
    // THEN
    assertions_1.Template.fromStack(stack).hasResourceProperties('AWS::Events::Rule', {
        ScheduleExpression: 'rate(1 minute)',
        State: 'ENABLED',
        Targets: [
            {
                Arn: {
                    'Fn::Join': [
                        '',
                        [
                            'arn:',
                            {
                                Ref: 'AWS::Partition',
                            },
                            ':logs:',
                            {
                                Ref: 'AWS::Region',
                            },
                            ':',
                            {
                                Ref: 'AWS::AccountId',
                            },
                            ':log-group:',
                            {
                                Ref: 'MyLogGroup5C0DAD85',
                            },
                        ],
                    ],
                },
                Id: 'Target0',
                InputTransformer: {
                    InputPathsMap: {
                        'time': '$.time',
                        'detail-type': '$.detail-type',
                    },
                    InputTemplate: '{"timestamp":<time>,"message":<detail-type>}',
                },
            },
        ],
    });
});
test('can use logEvent', () => {
    // GIVEN
    const stack = new cdk.Stack();
    const logGroup = new logs.LogGroup(stack, 'MyLogGroup', {
        logGroupName: '/aws/events/MyLogGroup',
    });
    const rule1 = new events.Rule(stack, 'Rule', {
        schedule: events.Schedule.rate(cdk.Duration.minutes(1)),
    });
    // WHEN
    rule1.addTarget(new targets.CloudWatchLogGroup(logGroup, {
        logEvent: lib_1.LogGroupTargetInput.fromObject({
            timestamp: events.EventField.time,
            message: events.EventField.fromPath('$'),
        }),
    }));
    // THEN
    assertions_1.Template.fromStack(stack).hasResourceProperties('AWS::Events::Rule', {
        ScheduleExpression: 'rate(1 minute)',
        State: 'ENABLED',
        Targets: [
            {
                Arn: {
                    'Fn::Join': [
                        '',
                        [
                            'arn:',
                            {
                                Ref: 'AWS::Partition',
                            },
                            ':logs:',
                            {
                                Ref: 'AWS::Region',
                            },
                            ':',
                            {
                                Ref: 'AWS::AccountId',
                            },
                            ':log-group:',
                            {
                                Ref: 'MyLogGroup5C0DAD85',
                            },
                        ],
                    ],
                },
                Id: 'Target0',
                InputTransformer: {
                    InputPathsMap: {
                        time: '$.time',
                        f2: '$',
                    },
                    InputTemplate: '{"timestamp":<time>,"message":<f2>}',
                },
            },
        ],
    });
});
(0, cdk_build_tools_1.testDeprecated)('specifying retry policy and dead letter queue', () => {
    // GIVEN
    const stack = new cdk.Stack();
    const logGroup = new logs.LogGroup(stack, 'MyLogGroup', {
        logGroupName: '/aws/events/MyLogGroup',
    });
    const rule1 = new events.Rule(stack, 'Rule', {
        schedule: events.Schedule.rate(cdk.Duration.minutes(1)),
    });
    const queue = new sqs.Queue(stack, 'Queue');
    // WHEN
    rule1.addTarget(new targets.CloudWatchLogGroup(logGroup, {
        event: events.RuleTargetInput.fromObject({
            timestamp: events.EventField.time,
            message: events.EventField.fromPath('$'),
        }),
        retryAttempts: 2,
        maxEventAge: cdk.Duration.hours(2),
        deadLetterQueue: queue,
    }));
    // THEN
    assertions_1.Template.fromStack(stack).hasResourceProperties('AWS::Events::Rule', {
        ScheduleExpression: 'rate(1 minute)',
        State: 'ENABLED',
        Targets: [
            {
                Arn: {
                    'Fn::Join': [
                        '',
                        [
                            'arn:',
                            {
                                Ref: 'AWS::Partition',
                            },
                            ':logs:',
                            {
                                Ref: 'AWS::Region',
                            },
                            ':',
                            {
                                Ref: 'AWS::AccountId',
                            },
                            ':log-group:',
                            {
                                Ref: 'MyLogGroup5C0DAD85',
                            },
                        ],
                    ],
                },
                DeadLetterConfig: {
                    Arn: {
                        'Fn::GetAtt': [
                            'Queue4A7E3555',
                            'Arn',
                        ],
                    },
                },
                Id: 'Target0',
                InputTransformer: {
                    InputPathsMap: {
                        time: '$.time',
                        f2: '$',
                    },
                    InputTemplate: '{"timestamp":<time>,"message":<f2>}',
                },
                RetryPolicy: {
                    MaximumEventAgeInSeconds: 7200,
                    MaximumRetryAttempts: 2,
                },
            },
        ],
    });
});
(0, cdk_build_tools_1.testDeprecated)('specifying retry policy with 0 retryAttempts', () => {
    // GIVEN
    const stack = new cdk.Stack();
    const logGroup = new logs.LogGroup(stack, 'MyLogGroup', {
        logGroupName: '/aws/events/MyLogGroup',
    });
    const rule1 = new events.Rule(stack, 'Rule', {
        schedule: events.Schedule.rate(cdk.Duration.minutes(1)),
    });
    // WHEN
    rule1.addTarget(new targets.CloudWatchLogGroup(logGroup, {
        event: events.RuleTargetInput.fromObject({
            timestamp: events.EventField.time,
            message: events.EventField.fromPath('$'),
        }),
        retryAttempts: 0,
    }));
    // THEN
    assertions_1.Template.fromStack(stack).hasResourceProperties('AWS::Events::Rule', {
        ScheduleExpression: 'rate(1 minute)',
        State: 'ENABLED',
        Targets: [
            {
                Arn: {
                    'Fn::Join': [
                        '',
                        [
                            'arn:',
                            {
                                Ref: 'AWS::Partition',
                            },
                            ':logs:',
                            {
                                Ref: 'AWS::Region',
                            },
                            ':',
                            {
                                Ref: 'AWS::AccountId',
                            },
                            ':log-group:',
                            {
                                Ref: 'MyLogGroup5C0DAD85',
                            },
                        ],
                    ],
                },
                Id: 'Target0',
                InputTransformer: {
                    InputPathsMap: {
                        time: '$.time',
                        f2: '$',
                    },
                    InputTemplate: '{"timestamp":<time>,"message":<f2>}',
                },
                RetryPolicy: {
                    MaximumRetryAttempts: 0,
                },
            },
        ],
    });
});
