// aws-cdk-lib/aws-docdb/test/parameter-group.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const assertions_1 = require("../../assertions");
const core_1 = require("../../core");
const lib_1 = require("../lib");
describe('ClusterParameterGroup', () => {
    test('check that instantiation works', () => {
        // GIVEN
        const stack = new core_1.Stack();
        // WHEN
        new lib_1.ClusterParameterGroup(stack, 'Params', {
            family: 'hello',
            description: 'desc',
            parameters: {
                key: 'value',
            },
        });
        // THEN
        assertions_1.Template.fromStack(stack).hasResourceProperties('AWS::DocDB::DBClusterParameterGroup', {
            Description: 'desc',
            Family: 'hello',
            Parameters: {
                key: 'value',
            },
        });
    });
    test('check automatically generated descriptions', () => {
        // GIVEN
        const stack = new core_1.Stack();
        // WHEN
        new lib_1.ClusterParameterGroup(stack, 'Params', {
            family: 'hello',
            parameters: {
                key: 'value',
            },
        });
        // THEN
        assertions_1.Template.fromStack(stack).hasResourceProperties('AWS::DocDB::DBClusterParameterGroup', {
            Description: 'Cluster parameter group for hello',
            Family: 'hello',
            Parameters: {
                key: 'value',
            },
        });
    });
});
