// aws-cdk-lib/aws-elasticloadbalancingv2-actions/test/cognito.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const assertions_1 = require("../../assertions");
const cognito = require("../../aws-cognito");
const ec2 = require("../../aws-ec2");
const elbv2 = require("../../aws-elasticloadbalancingv2");
const core_1 = require("../../core");
const actions = require("../lib");
test('Cognito Action', () => {
    // GIVEN
    const stack = new core_1.Stack();
    const vpc = new ec2.Vpc(stack, 'Stack');
    const lb = new elbv2.ApplicationLoadBalancer(stack, 'LB', { vpc });
    const userPool = new cognito.UserPool(stack, 'UserPool');
    const userPoolClient = new cognito.UserPoolClient(stack, 'Client', { userPool });
    const userPoolDomain = new cognito.UserPoolDomain(stack, 'Domain', {
        userPool,
        cognitoDomain: {
            domainPrefix: 'prefix',
        },
    });
    // WHEN
    lb.addListener('Listener', {
        port: 80,
        defaultAction: new actions.AuthenticateCognitoAction({
            userPool,
            userPoolClient,
            userPoolDomain,
            next: elbv2.ListenerAction.fixedResponse(200, {
                contentType: 'text/plain',
                messageBody: 'Authenticated',
            }),
        }),
    });
    // THEN
    assertions_1.Template.fromStack(stack).hasResourceProperties('AWS::ElasticLoadBalancingV2::Listener', {
        DefaultActions: [
            {
                AuthenticateCognitoConfig: {
                    UserPoolArn: { 'Fn::GetAtt': ['UserPool6BA7E5F2', 'Arn'] },
                    UserPoolClientId: { Ref: 'Client4A7F64DF' },
                    UserPoolDomain: { Ref: 'Domain66AC69E0' },
                },
                Order: 1,
                Type: 'authenticate-cognito',
            },
            {
                FixedResponseConfig: {
                    ContentType: 'text/plain',
                    MessageBody: 'Authenticated',
                    StatusCode: '200',
                },
                Order: 2,
                Type: 'fixed-response',
            },
        ],
    });
});
