// serverless-lift/dist/src/constructs/aws/DatabaseDynamoDBSingleTable.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var DatabaseDynamoDBSingleTable_exports = {};
__export(DatabaseDynamoDBSingleTable_exports, {
  DatabaseDynamoDBSingleTable: () => DatabaseDynamoDBSingleTable
});
module.exports = __toCommonJS(DatabaseDynamoDBSingleTable_exports);
var import_aws_cdk_lib = require("aws-cdk-lib");
var import_aws_dynamodb = require("aws-cdk-lib/aws-dynamodb");
var import_abstracts = require("../abstracts");
var import_CloudFormation = require("../../CloudFormation");
const DATABASE_DEFINITION = {
  type: "object",
  properties: {
    type: { const: "database/dynamodb-single-table" },
    localSecondaryIndexes: { type: "boolean" },
    gsiCount: { type: "integer", minimum: 1, maximum: 20 }
  },
  additionalProperties: false
};
const DATABASE_DEFAULTS = {
  type: "database/dynamodb-single-table",
  localSecondaryIndexes: false,
  gsiCount: 0
};
class DatabaseDynamoDBSingleTable extends import_abstracts.AwsConstruct {
  constructor(scope, id, configuration, provider) {
    super(scope, id);
    this.provider = provider;
    const resolvedConfiguration = Object.assign({}, DATABASE_DEFAULTS, configuration);
    this.table = new import_aws_dynamodb.Table(this, "Table", {
      partitionKey: { name: "PK", type: import_aws_dynamodb.AttributeType.STRING },
      sortKey: { name: "SK", type: import_aws_dynamodb.AttributeType.STRING },
      billingMode: import_aws_dynamodb.BillingMode.PAY_PER_REQUEST,
      pointInTimeRecovery: true,
      timeToLiveAttribute: "TimeToLive",
      stream: import_aws_dynamodb.StreamViewType.NEW_AND_OLD_IMAGES
    });
    if (resolvedConfiguration.localSecondaryIndexes) {
      for (let localSecondaryIndex = 1; localSecondaryIndex <= 5; localSecondaryIndex++) {
        this.table.addLocalSecondaryIndex({
          indexName: `LSI-${localSecondaryIndex}`,
          sortKey: { name: `LSI-${localSecondaryIndex}-SK`, type: import_aws_dynamodb.AttributeType.STRING }
        });
      }
    }
    if (resolvedConfiguration.gsiCount > 0) {
      for (let globalSecondaryIndex = 1; globalSecondaryIndex <= resolvedConfiguration.gsiCount; globalSecondaryIndex++) {
        this.table.addGlobalSecondaryIndex({
          indexName: `GSI-${globalSecondaryIndex}`,
          partitionKey: { name: `GSI-${globalSecondaryIndex}-PK`, type: import_aws_dynamodb.AttributeType.STRING },
          sortKey: { name: `GSI-${globalSecondaryIndex}-SK`, type: import_aws_dynamodb.AttributeType.STRING }
        });
      }
    }
    this.tableNameOutput = new import_aws_cdk_lib.CfnOutput(this, "TableName", {
      value: this.table.tableName
    });
  }
  permissions() {
    return [
      new import_CloudFormation.PolicyStatement(
        [
          "dynamodb:GetItem",
          "dynamodb:BatchGetItem",
          "dynamodb:Query",
          "dynamodb:Scan",
          "dynamodb:PutItem",
          "dynamodb:DeleteItem",
          "dynamodb:BatchWriteItem",
          "dynamodb:UpdateItem"
        ],
        [this.table.tableArn, import_aws_cdk_lib.Stack.of(this).resolve(import_aws_cdk_lib.Fn.join("/", [this.table.tableArn, "index", "*"]))]
      )
    ];
  }
  outputs() {
    return {
      tableName: () => this.getTableName()
    };
  }
  variables() {
    return {
      tableName: this.table.tableName,
      tableArn: this.table.tableArn,
      tableStreamArn: this.table.tableStreamArn
    };
  }
  extend() {
    return {
      table: this.table.node.defaultChild
    };
  }
  async getTableName() {
    return this.provider.getStackOutput(this.tableNameOutput);
  }
}
DatabaseDynamoDBSingleTable.type = "database/dynamodb-single-table";
DatabaseDynamoDBSingleTable.schema = DATABASE_DEFINITION;
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  DatabaseDynamoDBSingleTable
});
