// serverless-lift/dist/src/constructs/aws/SinglePageApp.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var SinglePageApp_exports = {};
__export(SinglePageApp_exports, {
  SinglePageApp: () => SinglePageApp
});
module.exports = __toCommonJS(SinglePageApp_exports);
var cloudfront = __toESM(require("aws-cdk-lib/aws-cloudfront"));
var import_aws_cloudfront = require("aws-cdk-lib/aws-cloudfront");
var import_cloudfrontFunctions = require("../../classes/cloudfrontFunctions");
var import_getDefaultCfnFunctionAssociations = require("../../utils/getDefaultCfnFunctionAssociations");
var import_naming = require("../../utils/naming");
var import_StaticWebsiteAbstract = require("./abstracts/StaticWebsiteAbstract");
class SinglePageApp extends import_StaticWebsiteAbstract.StaticWebsiteAbstract {
  constructor(scope, id, configuration, provider) {
    super(scope, id, configuration, provider);
    this.id = id;
    this.configuration = configuration;
    this.provider = provider;
    const cfnDistribution = this.distribution.node.defaultChild;
    const requestFunction = this.createRequestFunction();
    const defaultBehaviorFunctionAssociations = (0, import_getDefaultCfnFunctionAssociations.getCfnFunctionAssociations)(cfnDistribution);
    cfnDistribution.addOverride("Properties.DistributionConfig.DefaultCacheBehavior.FunctionAssociations", [
      ...defaultBehaviorFunctionAssociations,
      { EventType: import_aws_cloudfront.FunctionEventType.VIEWER_REQUEST, FunctionARN: requestFunction.functionArn }
    ]);
  }
  createRequestFunction() {
    let additionalCode = "";
    if (this.configuration.redirectToMainDomain === true) {
      additionalCode += (0, import_cloudfrontFunctions.redirectToMainDomain)(this.domains);
    }
    const code = `var REDIRECT_REGEX = /^[^.]+$|\\.(?!(css|gif|ico|jpg|jpeg|js|png|txt|svg|woff|woff2|ttf|map|json|webp|xml|pdf|webmanifest|avif)$)([^.]+$)/;

function handler(event) {
    var uri = event.request.uri;
    var request = event.request;
    var isUriToRedirect = REDIRECT_REGEX.test(uri);

    if (isUriToRedirect) {
        request.uri = "/index.html";
    }${additionalCode}

    return event.request;
}`;
    const functionName = (0, import_naming.ensureNameMaxLength)(
      `${this.provider.stackName}-${this.provider.region}-${this.id}-request`,
      64
    );
    return new cloudfront.Function(this, "RequestFunction", {
      functionName,
      code: cloudfront.FunctionCode.fromInline(code)
    });
  }
}
SinglePageApp.type = "single-page-app";
SinglePageApp.schema = import_StaticWebsiteAbstract.COMMON_STATIC_WEBSITE_DEFINITION;
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  SinglePageApp
});
