// aws-cdk-lib/aws-apigateway/test/util.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const lib_1 = require("../lib");
const util_1 = require("../lib/util");
describe('util', () => {
    describe('parseMethodResourcePath', () => {
        test('fails if path does not start with a /', () => {
            expect(() => (0, util_1.parseMethodOptionsPath)('foo')).toThrow(/Method options path must start with \'\/\'/);
        });
        test('fails if there are less than two components', () => {
            expect(() => (0, util_1.parseMethodOptionsPath)('/')).toThrow(/Method options path must include at least two components/);
            expect(() => (0, util_1.parseMethodOptionsPath)('/foo')).toThrow(/Method options path must include at least two components/);
            expect(() => (0, util_1.parseMethodOptionsPath)('/foo/')).toThrow(/Invalid HTTP method ""/);
        });
        test('fails if a non-supported http method is used', () => {
            expect(() => (0, util_1.parseMethodOptionsPath)('/foo/bar')).toThrow(/Invalid HTTP method "BAR"/);
        });
        test('extracts resource path and method correctly', () => {
            expect((0, util_1.parseMethodOptionsPath)('/foo/GET')).toEqual({ resourcePath: '/~1foo', httpMethod: 'GET' });
            expect((0, util_1.parseMethodOptionsPath)('/foo/bar/GET')).toEqual({ resourcePath: '/~1foo~1bar', httpMethod: 'GET' });
            expect((0, util_1.parseMethodOptionsPath)('/foo/*/GET')).toEqual({ resourcePath: '/~1foo~1*', httpMethod: 'GET' });
            expect((0, util_1.parseMethodOptionsPath)('/*/GET')).toEqual({ resourcePath: '/*', httpMethod: 'GET' });
            expect((0, util_1.parseMethodOptionsPath)('/*/*')).toEqual({ resourcePath: '/*', httpMethod: '*' });
            expect((0, util_1.parseMethodOptionsPath)('//POST')).toEqual({ resourcePath: '/', httpMethod: 'POST' });
        });
    });
    describe('parseAwsApiCall', () => {
        test('fails if "actionParams" is set but "action" is undefined', () => {
            expect(() => (0, util_1.parseAwsApiCall)(undefined, undefined, { foo: '123' })).toThrow(/"actionParams" requires that "action" will be set/);
        });
        test('fails since "action" and "path" are mutually exclusive', () => {
            expect(() => (0, util_1.parseAwsApiCall)('foo', 'bar')).toThrow(/"path" and "action" are mutually exclusive \(path="foo", action="bar"\)/);
        });
        test('fails if "path" and "action" are both undefined', () => {
            expect(() => (0, util_1.parseAwsApiCall)()).toThrow(/Either "path" or "action" are required/);
        });
        test('"path" mode', () => {
            expect((0, util_1.parseAwsApiCall)('my/path')).toEqual({ apiType: 'path', apiValue: 'my/path' });
        });
        test('"action" mode with no parameters', () => {
            expect((0, util_1.parseAwsApiCall)(undefined, 'MyAction')).toEqual({ apiType: 'action', apiValue: 'MyAction' });
        });
        test('"action" mode with parameters (url-encoded)', () => {
            expect((0, util_1.parseAwsApiCall)(undefined, 'GetObject', { Bucket: 'MyBucket', Key: 'MyKey' })).toEqual({
                apiType: 'action',
                apiValue: 'GetObject&Bucket=MyBucket&Key=MyKey',
            });
        });
    });
    describe('JsonSchemaMapper.toCfnJsonSchema', () => {
        test('maps "ref" found under properties', () => {
            const schema = {
                type: lib_1.JsonSchemaType.OBJECT,
                properties: {
                    collection: {
                        type: lib_1.JsonSchemaType.ARRAY,
                        items: {
                            ref: '#/some/reference',
                        },
                        uniqueItems: true,
                    },
                },
                required: ['collection'],
            };
            const actual = util_1.JsonSchemaMapper.toCfnJsonSchema(schema);
            expect(actual).toEqual({
                $schema: 'http://json-schema.org/draft-04/schema#',
                type: 'object',
                properties: {
                    collection: {
                        type: 'array',
                        items: {
                            $ref: '#/some/reference',
                        },
                        uniqueItems: true,
                    },
                },
                required: ['collection'],
            });
        });
        test('does not map a "ref" property name', () => {
            const schema = {
                type: lib_1.JsonSchemaType.OBJECT,
                properties: {
                    ref: {
                        type: lib_1.JsonSchemaType.ARRAY,
                        items: {
                            ref: '#/some/reference',
                        },
                        uniqueItems: true,
                    },
                },
                required: ['ref'],
            };
            const actual = util_1.JsonSchemaMapper.toCfnJsonSchema(schema);
            expect(actual).toEqual({
                $schema: 'http://json-schema.org/draft-04/schema#',
                type: 'object',
                properties: {
                    ref: {
                        type: 'array',
                        items: {
                            $ref: '#/some/reference',
                        },
                        uniqueItems: true,
                    },
                },
                required: ['ref'],
            });
        });
        test('"default" for enum', () => {
            const schema = {
                type: lib_1.JsonSchemaType.STRING,
                enum: ['green', 'blue', 'red'],
                default: 'blue',
            };
            const actual = util_1.JsonSchemaMapper.toCfnJsonSchema(schema);
            expect(actual).toEqual({
                $schema: 'http://json-schema.org/draft-04/schema#',
                type: 'string',
                enum: ['green', 'blue', 'red'],
                default: 'blue',
            });
        });
        test('"id" maps to "id" when using DRAFT-04', () => {
            const schema = {
                schema: lib_1.JsonSchemaVersion.DRAFT4,
                id: 'http://json-schema.org/draft-04/schema#',
            };
            const actual = util_1.JsonSchemaMapper.toCfnJsonSchema(schema);
            expect(actual).toEqual({
                $schema: 'http://json-schema.org/draft-04/schema#',
                id: 'http://json-schema.org/draft-04/schema#',
            });
        });
    });
});
