// aws-cdk-lib/aws-cloudfront-origins/test/http-origin.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const cloudfront = require("../../aws-cloudfront");
const core_1 = require("../../core");
const lib_1 = require("../lib");
let app;
let stack;
beforeEach(() => {
    app = new core_1.App();
    new core_1.Stack(app, 'Stack', {
        env: { account: '1234', region: 'testregion' },
    });
});
test('Renders minimal example with just a domain name', () => {
    const origin = new lib_1.HttpOrigin('www.example.com');
    const originBindConfig = origin.bind(stack, { originId: 'StackOrigin029E19582' });
    expect(originBindConfig.originProperty).toEqual({
        id: 'StackOrigin029E19582',
        domainName: 'www.example.com',
        originCustomHeaders: undefined,
        originPath: undefined,
        customOriginConfig: {
            originProtocolPolicy: 'https-only',
            originSslProtocols: [
                'TLSv1.2',
            ],
        },
    });
});
test('renders an example with all available props', () => {
    const origin = new lib_1.HttpOrigin('www.example.com', {
        originId: 'MyCustomOrigin',
        originPath: '/app',
        connectionTimeout: core_1.Duration.seconds(5),
        connectionAttempts: 2,
        customHeaders: { AUTH: 'NONE' },
        protocolPolicy: cloudfront.OriginProtocolPolicy.MATCH_VIEWER,
        httpPort: 8080,
        httpsPort: 8443,
        readTimeout: core_1.Duration.seconds(45),
        keepaliveTimeout: core_1.Duration.seconds(3),
        originSslProtocols: [cloudfront.OriginSslPolicy.TLS_V1_2],
    });
    const originBindConfig = origin.bind(stack, { originId: 'StackOrigin029E19582' });
    expect(originBindConfig.originProperty).toEqual({
        id: 'MyCustomOrigin',
        domainName: 'www.example.com',
        originPath: '/app',
        connectionTimeout: 5,
        connectionAttempts: 2,
        originCustomHeaders: [{
                headerName: 'AUTH',
                headerValue: 'NONE',
            }],
        customOriginConfig: {
            originProtocolPolicy: 'match-viewer',
            originSslProtocols: [
                'TLSv1.2',
            ],
            httpPort: 8080,
            httpsPort: 8443,
            originReadTimeout: 45,
            originKeepaliveTimeout: 3,
        },
    });
});
test.each([
    core_1.Duration.seconds(0),
    core_1.Duration.seconds(181),
    core_1.Duration.minutes(5),
])('validates readTimeout is an integer between 1 and 180 seconds - out of bounds', (readTimeout) => {
    expect(() => {
        new lib_1.HttpOrigin('www.example.com', {
            readTimeout,
        });
    }).toThrow(`readTimeout: Must be an int between 1 and 180 seconds (inclusive); received ${readTimeout.toSeconds()}.`);
});
test.each([
    core_1.Duration.seconds(0.5),
    core_1.Duration.seconds(60.5),
])('validates readTimeout is an integer between 1 and 180 seconds - not an int', (readTimeout) => {
    expect(() => {
        new lib_1.HttpOrigin('www.example.com', {
            readTimeout,
        });
    }).toThrow(/must be a whole number of/);
});
test.each([
    core_1.Duration.seconds(0),
    core_1.Duration.seconds(181),
    core_1.Duration.minutes(5),
])('validates keepaliveTimeout is an integer between 1 and 180 seconds - out of bounds', (keepaliveTimeout) => {
    expect(() => {
        new lib_1.HttpOrigin('www.example.com', {
            keepaliveTimeout,
        });
    }).toThrow(`keepaliveTimeout: Must be an int between 1 and 180 seconds (inclusive); received ${keepaliveTimeout.toSeconds()}.`);
});
test.each([
    core_1.Duration.seconds(0.5),
    core_1.Duration.seconds(60.5),
])('validates keepaliveTimeout is an integer between 1 and 180 seconds - not an int', (keepaliveTimeout) => {
    expect(() => {
        new lib_1.HttpOrigin('www.example.com', {
            keepaliveTimeout,
        });
    }).toThrow(/must be a whole number of/);
});
