// aws-cdk-lib/aws-cloudfront-origins/test/load-balancer-origin.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const cloudfront = require("../../aws-cloudfront");
const elbv2 = require("../../aws-elasticloadbalancingv2");
const core_1 = require("../../core");
const lib_1 = require("../lib");
let app;
let stack;
beforeEach(() => {
    app = new core_1.App();
    stack = new core_1.Stack(app, 'Stack', {
        env: { account: '1234', region: 'testregion' },
    });
});
test('Renders minimal example with just a load balancer', () => {
    const loadBalancer = elbv2.NetworkLoadBalancer.fromNetworkLoadBalancerAttributes(stack, 'elbv2', {
        loadBalancerArn: 'arn:aws:elasticloadbalancing:us-west-2:111111111111:loadbalancer/net/mylb/5d1b75f4f1cee11e',
        loadBalancerDnsName: 'mylb-5d1b75f4f1cee11e.elb.us-west-2.amazonaws.com',
    });
    const origin = new lib_1.LoadBalancerV2Origin(loadBalancer);
    const originBindConfig = origin.bind(stack, { originId: 'StackOrigin029E19582' });
    expect(originBindConfig.originProperty).toEqual({
        id: 'StackOrigin029E19582',
        domainName: loadBalancer.loadBalancerDnsName,
        customOriginConfig: {
            originProtocolPolicy: 'https-only',
            originSslProtocols: [
                'TLSv1.2',
            ],
        },
    });
});
test('Can customize properties of the origin', () => {
    const loadBalancer = elbv2.NetworkLoadBalancer.fromNetworkLoadBalancerAttributes(stack, 'elbv2', {
        loadBalancerArn: 'arn:aws:elasticloadbalancing:us-west-2:111111111111:loadbalancer/net/mylb/5d1b75f4f1cee11e',
        loadBalancerDnsName: 'mylb-5d1b75f4f1cee11e.elb.us-west-2.amazonaws.com',
    });
    const origin = new lib_1.LoadBalancerV2Origin(loadBalancer, {
        connectionAttempts: 3,
        connectionTimeout: core_1.Duration.seconds(5),
        protocolPolicy: cloudfront.OriginProtocolPolicy.MATCH_VIEWER,
        originId: 'MyCustomOrigin',
    });
    const originBindConfig = origin.bind(stack, { originId: 'StackOrigin029E19582' });
    expect(originBindConfig.originProperty).toEqual({
        id: 'MyCustomOrigin',
        domainName: loadBalancer.loadBalancerDnsName,
        connectionAttempts: 3,
        connectionTimeout: 5,
        customOriginConfig: {
            originProtocolPolicy: 'match-viewer',
            originSslProtocols: [
                'TLSv1.2',
            ],
        },
    });
});
