// aws-cdk-lib/aws-ecs-patterns/test/ec2/l3s-v2.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const assertions_1 = require("../../../assertions");
const aws_autoscaling_1 = require("../../../aws-autoscaling");
const aws_certificatemanager_1 = require("../../../aws-certificatemanager");
const ec2 = require("../../../aws-ec2");
const aws_ec2_1 = require("../../../aws-ec2");
const aws_ecs_1 = require("../../../aws-ecs");
const aws_elasticloadbalancingv2_1 = require("../../../aws-elasticloadbalancingv2");
const aws_iam_1 = require("../../../aws-iam");
const aws_route53_1 = require("../../../aws-route53");
const aws_servicediscovery_1 = require("../../../aws-servicediscovery");
const core_1 = require("../../../core");
const lib_1 = require("../../lib");
describe('When Application Load Balancer', () => {
    test('test ECS ALB construct with default settings', () => {
        // GIVEN
        const stack = new core_1.Stack();
        const vpc = new aws_ec2_1.Vpc(stack, 'VPC');
        const cluster = new aws_ecs_1.Cluster(stack, 'Cluster', { vpc });
        cluster.addAsgCapacityProvider(new aws_ecs_1.AsgCapacityProvider(stack, 'DefaultAutoScalingGroupProvider', {
            autoScalingGroup: new aws_autoscaling_1.AutoScalingGroup(stack, 'DefaultAutoScalingGroup', {
                vpc,
                instanceType: new ec2.InstanceType('t2.micro'),
                machineImage: aws_ec2_1.MachineImage.latestAmazonLinux(),
            }),
        }));
        // WHEN
        new lib_1.ApplicationMultipleTargetGroupsEc2Service(stack, 'Service', {
            cluster,
            memoryLimitMiB: 1024,
            taskImageOptions: {
                image: aws_ecs_1.ContainerImage.fromRegistry('test'),
            },
        });
        // THEN - stack contains a load balancer, a service, and a target group.
        assertions_1.Template.fromStack(stack).resourceCountIs('AWS::ElasticLoadBalancingV2::LoadBalancer', 1);
        assertions_1.Template.fromStack(stack).hasResourceProperties('AWS::ECS::Service', {
            LaunchType: 'EC2',
        });
        assertions_1.Template.fromStack(stack).hasResourceProperties('AWS::ECS::TaskDefinition', {
            ContainerDefinitions: [
                assertions_1.Match.objectLike({
                    Image: 'test',
                    LogConfiguration: {
                        LogDriver: 'awslogs',
                        Options: {
                            'awslogs-group': {
                                Ref: 'ServiceTaskDefwebLogGroup2A898F61',
                            },
                            'awslogs-stream-prefix': 'Service',
                            'awslogs-region': {
                                Ref: 'AWS::Region',
                            },
                        },
                    },
                    Memory: 1024,
                    Name: 'web',
                    PortMappings: [
                        {
                            ContainerPort: 80,
                            HostPort: 0,
                            Protocol: 'tcp',
                        },
                    ],
                }),
            ],
            NetworkMode: 'bridge',
            RequiresCompatibilities: [
                'EC2',
            ],
        });
    });
    test('test ECS ALB construct with all settings', () => {
        // GIVEN
        const stack = new core_1.Stack();
        const vpc = new aws_ec2_1.Vpc(stack, 'VPC');
        const cluster = new aws_ecs_1.Cluster(stack, 'Cluster', { vpc });
        cluster.addAsgCapacityProvider(new aws_ecs_1.AsgCapacityProvider(stack, 'DefaultAutoScalingGroupProvider', {
            autoScalingGroup: new aws_autoscaling_1.AutoScalingGroup(stack, 'DefaultAutoScalingGroup', {
                vpc,
                instanceType: new ec2.InstanceType('t2.micro'),
                machineImage: aws_ec2_1.MachineImage.latestAmazonLinux(),
            }),
        }));
        const zone = new aws_route53_1.PublicHostedZone(stack, 'HostedZone', { zoneName: 'example.com' });
        // WHEN
        new lib_1.ApplicationMultipleTargetGroupsEc2Service(stack, 'Service', {
            cluster,
            memoryLimitMiB: 1024,
            taskImageOptions: {
                image: aws_ecs_1.ContainerImage.fromRegistry('test'),
                containerName: 'myContainer',
                containerPorts: [80, 90],
                enableLogging: false,
                environment: {
                    TEST_ENVIRONMENT_VARIABLE1: 'test environment variable 1 value',
                    TEST_ENVIRONMENT_VARIABLE2: 'test environment variable 2 value',
                },
                logDriver: new aws_ecs_1.AwsLogDriver({
                    streamPrefix: 'TestStream',
                }),
                family: 'Ec2TaskDef',
                executionRole: new aws_iam_1.Role(stack, 'ExecutionRole', {
                    path: '/',
                    assumedBy: new aws_iam_1.CompositePrincipal(new aws_iam_1.ServicePrincipal('ecs.amazonaws.com'), new aws_iam_1.ServicePrincipal('ecs-tasks.amazonaws.com')),
                }),
                taskRole: new aws_iam_1.Role(stack, 'TaskRole', {
                    assumedBy: new aws_iam_1.ServicePrincipal('ecs-tasks.amazonaws.com'),
                }),
                dockerLabels: { label1: 'labelValue1', label2: 'labelValue2' },
            },
            cpu: 256,
            desiredCount: 3,
            enableECSManagedTags: true,
            enableExecuteCommand: true,
            healthCheckGracePeriod: core_1.Duration.millis(2000),
            loadBalancers: [
                {
                    name: 'lb',
                    domainName: 'api.example.com',
                    domainZone: zone,
                    publicLoadBalancer: false,
                    listeners: [
                        {
                            name: 'listener',
                            protocol: aws_elasticloadbalancingv2_1.ApplicationProtocol.HTTPS,
                            certificate: aws_certificatemanager_1.Certificate.fromCertificateArn(stack, 'Cert', 'helloworld'),
                            sslPolicy: aws_elasticloadbalancingv2_1.SslPolicy.TLS12_EXT,
                        },
                    ],
                },
            ],
            propagateTags: aws_ecs_1.PropagatedTagSource.SERVICE,
            memoryReservationMiB: 1024,
            serviceName: 'myService',
            targetGroups: [
                {
                    containerPort: 80,
                    listener: 'listener',
                },
                {
                    containerPort: 90,
                    listener: 'listener',
                    pathPattern: 'a/b/c',
                    priority: 10,
                    protocol: aws_ecs_1.Protocol.TCP,
                },
            ],
            placementStrategies: [aws_ecs_1.PlacementStrategy.spreadAcrossInstances(), aws_ecs_1.PlacementStrategy.packedByCpu(), aws_ecs_1.PlacementStrategy.randomly()],
            placementConstraints: [aws_ecs_1.PlacementConstraint.memberOf('attribute:ecs.instance-type =~ m5a.*')],
        });
        // THEN
        assertions_1.Template.fromStack(stack).hasResourceProperties('AWS::ECS::Service', {
            DesiredCount: 3,
            LaunchType: 'EC2',
            EnableECSManagedTags: true,
            EnableExecuteCommand: true,
            HealthCheckGracePeriodSeconds: 2,
            LoadBalancers: [
                {
                    ContainerName: 'myContainer',
                    ContainerPort: 80,
                    TargetGroupArn: {
                        Ref: 'ServicelblistenerECSTargetGroupmyContainer80GroupAD83584A',
                    },
                },
                {
                    ContainerName: 'myContainer',
                    ContainerPort: 90,
                    TargetGroupArn: {
                        Ref: 'ServicelblistenerECSTargetGroupmyContainer90GroupF5A6D3A0',
                    },
                },
            ],
            PropagateTags: 'SERVICE',
            ServiceName: 'myService',
            PlacementConstraints: [{ Type: 'memberOf', Expression: 'attribute:ecs.instance-type =~ m5a.*' }],
            PlacementStrategies: [{ Field: 'instanceId', Type: 'spread' }, { Field: 'CPU', Type: 'binpack' }, { Type: 'random' }],
        });
        assertions_1.Template.fromStack(stack).hasResourceProperties('AWS::ECS::TaskDefinition', {
            ContainerDefinitions: [
                {
                    Cpu: 256,
                    Environment: [
                        {
                            Name: 'TEST_ENVIRONMENT_VARIABLE1',
                            Value: 'test environment variable 1 value',
                        },
                        {
                            Name: 'TEST_ENVIRONMENT_VARIABLE2',
                            Value: 'test environment variable 2 value',
                        },
                    ],
                    Essential: true,
                    Image: 'test',
                    LogConfiguration: {
                        LogDriver: 'awslogs',
                        Options: {
                            'awslogs-group': {
                                Ref: 'ServiceTaskDefmyContainerLogGroup0A87368B',
                            },
                            'awslogs-stream-prefix': 'TestStream',
                            'awslogs-region': {
                                Ref: 'AWS::Region',
                            },
                        },
                    },
                    Memory: 1024,
                    MemoryReservation: 1024,
                    Name: 'myContainer',
                    PortMappings: [
                        {
                            ContainerPort: 80,
                            HostPort: 0,
                            Protocol: 'tcp',
                        },
                        {
                            ContainerPort: 90,
                            HostPort: 0,
                            Protocol: 'tcp',
                        },
                    ],
                    DockerLabels: {
                        label1: 'labelValue1',
                        label2: 'labelValue2',
                    },
                },
            ],
            ExecutionRoleArn: {
                'Fn::GetAtt': [
                    'ExecutionRole605A040B',
                    'Arn',
                ],
            },
            Family: 'ServiceTaskDef79D79521',
            NetworkMode: 'bridge',
            RequiresCompatibilities: [
                'EC2',
            ],
            TaskRoleArn: {
                'Fn::GetAtt': [
                    'TaskRole30FC0FBB',
                    'Arn',
                ],
            },
        });
        assertions_1.Template.fromStack(stack).hasResourceProperties('AWS::ElasticLoadBalancingV2::Listener', {
            Port: 443,
            Protocol: 'HTTPS',
            Certificates: [{
                    CertificateArn: 'helloworld',
                }],
            SslPolicy: aws_elasticloadbalancingv2_1.SslPolicy.TLS12_EXT,
        });
    });
    test('able to pass pre-defined task definition', () => {
        // GIVEN
        const stack = new core_1.Stack();
        const vpc = new aws_ec2_1.Vpc(stack, 'VPC');
        const cluster = new aws_ecs_1.Cluster(stack, 'Cluster', { vpc });
        cluster.addAsgCapacityProvider(new aws_ecs_1.AsgCapacityProvider(stack, 'DefaultAutoScalingGroupProvider', {
            autoScalingGroup: new aws_autoscaling_1.AutoScalingGroup(stack, 'DefaultAutoScalingGroup', {
                vpc,
                instanceType: new ec2.InstanceType('t2.micro'),
                machineImage: aws_ec2_1.MachineImage.latestAmazonLinux(),
            }),
        }));
        const taskDefinition = new aws_ecs_1.Ec2TaskDefinition(stack, 'Ec2TaskDef');
        const container = taskDefinition.addContainer('web', {
            image: aws_ecs_1.ContainerImage.fromRegistry('amazon/amazon-ecs-sample'),
            memoryLimitMiB: 512,
        });
        container.addPortMappings({
            containerPort: 80,
        });
        // WHEN
        new lib_1.ApplicationMultipleTargetGroupsEc2Service(stack, 'Service', {
            cluster,
            taskDefinition,
        });
        // THEN
        assertions_1.Template.fromStack(stack).hasResourceProperties('AWS::ECS::TaskDefinition', {
            ContainerDefinitions: [
                {
                    Essential: true,
                    Image: 'amazon/amazon-ecs-sample',
                    Memory: 512,
                    Name: 'web',
                    PortMappings: [
                        {
                            ContainerPort: 80,
                            HostPort: 0,
                            Protocol: 'tcp',
                        },
                    ],
                },
            ],
            Family: 'Ec2TaskDef',
            NetworkMode: 'bridge',
            RequiresCompatibilities: [
                'EC2',
            ],
        });
    });
    test('able to output correct load balancer DNS and URLs for each protocol type', () => {
        // GIVEN
        const stack = new core_1.Stack();
        const vpc = new aws_ec2_1.Vpc(stack, 'VPC');
        const cluster = new aws_ecs_1.Cluster(stack, 'Cluster', { vpc });
        cluster.addAsgCapacityProvider(new aws_ecs_1.AsgCapacityProvider(stack, 'DefaultAutoScalingGroupProvider', {
            autoScalingGroup: new aws_autoscaling_1.AutoScalingGroup(stack, 'DefaultAutoScalingGroup', {
                vpc,
                instanceType: new ec2.InstanceType('t2.micro'),
                machineImage: aws_ec2_1.MachineImage.latestAmazonLinux(),
            }),
        }));
        const zone = new aws_route53_1.PublicHostedZone(stack, 'HostedZone', { zoneName: 'example.com' });
        // WHEN
        new lib_1.ApplicationMultipleTargetGroupsEc2Service(stack, 'Service', {
            cluster,
            memoryLimitMiB: 1024,
            taskImageOptions: {
                image: aws_ecs_1.ContainerImage.fromRegistry('test'),
            },
            loadBalancers: [
                {
                    name: 'lb1',
                    domainName: 'api.example.com',
                    domainZone: zone,
                    listeners: [
                        {
                            name: 'listener1',
                            protocol: aws_elasticloadbalancingv2_1.ApplicationProtocol.HTTPS,
                            certificate: aws_certificatemanager_1.Certificate.fromCertificateArn(stack, 'Cert', 'helloworld'),
                        },
                        {
                            name: 'listener2',
                            protocol: aws_elasticloadbalancingv2_1.ApplicationProtocol.HTTP,
                        },
                    ],
                },
                {
                    name: 'lb3',
                    listeners: [
                        {
                            name: 'listener3',
                            protocol: aws_elasticloadbalancingv2_1.ApplicationProtocol.HTTP,
                        },
                    ],
                },
            ],
            targetGroups: [
                {
                    containerPort: 80,
                    listener: 'listener1',
                },
                {
                    containerPort: 90,
                    listener: 'listener2',
                },
                {
                    containerPort: 70,
                    listener: 'listener3',
                },
            ],
        });
        // THEN
        const outputs = assertions_1.Template.fromStack(stack).findOutputs('*');
        expect(outputs).toEqual({
            ServiceLoadBalancerDNSlb175E78BFE: {
                Value: {
                    'Fn::GetAtt': [
                        'Servicelb152C7F4F9',
                        'DNSName',
                    ],
                },
            },
            ServiceServiceURLlb1https5C0C4079: {
                Value: {
                    'Fn::Join': [
                        '',
                        [
                            'https://',
                            {
                                Ref: 'ServiceDNSlb12BA1FAD3',
                            },
                        ],
                    ],
                },
            },
            ServiceServiceURLlb1http65F0546A: {
                Value: {
                    'Fn::Join': [
                        '',
                        [
                            'http://',
                            {
                                Ref: 'ServiceDNSlb12BA1FAD3',
                            },
                        ],
                    ],
                },
            },
            ServiceLoadBalancerDNSlb32F273F27: {
                Value: {
                    'Fn::GetAtt': [
                        'Servicelb3A583D5E7',
                        'DNSName',
                    ],
                },
            },
            ServiceServiceURLlb3http40F9CADC: {
                Value: {
                    'Fn::Join': [
                        '',
                        [
                            'http://',
                            {
                                'Fn::GetAtt': [
                                    'Servicelb3A583D5E7',
                                    'DNSName',
                                ],
                            },
                        ],
                    ],
                },
            },
        });
    });
    test('errors if no essential container in pre-defined task definition', () => {
        // GIVEN
        const stack = new core_1.Stack();
        const vpc = new aws_ec2_1.Vpc(stack, 'VPC');
        const cluster = new aws_ecs_1.Cluster(stack, 'Cluster', { vpc });
        cluster.addAsgCapacityProvider(new aws_ecs_1.AsgCapacityProvider(stack, 'DefaultAutoScalingGroupProvider', {
            autoScalingGroup: new aws_autoscaling_1.AutoScalingGroup(stack, 'DefaultAutoScalingGroup', {
                vpc,
                instanceType: new ec2.InstanceType('t2.micro'),
                machineImage: aws_ec2_1.MachineImage.latestAmazonLinux(),
            }),
        }));
        const taskDefinition = new aws_ecs_1.Ec2TaskDefinition(stack, 'Ec2TaskDef');
        // THEN
        expect(() => {
            new lib_1.ApplicationMultipleTargetGroupsEc2Service(stack, 'Service', {
                cluster,
                taskDefinition,
            });
        }).toThrow(/At least one essential container must be specified/);
    });
    test('set default load balancer, listener, target group correctly', () => {
        // GIVEN
        const stack = new core_1.Stack();
        const vpc = new aws_ec2_1.Vpc(stack, 'VPC');
        const zone = new aws_route53_1.PublicHostedZone(stack, 'HostedZone', { zoneName: 'example.com' });
        // WHEN
        const ecsService = new lib_1.ApplicationMultipleTargetGroupsEc2Service(stack, 'Service', {
            vpc,
            memoryLimitMiB: 1024,
            taskImageOptions: {
                image: aws_ecs_1.ContainerImage.fromRegistry('test'),
            },
            loadBalancers: [
                {
                    name: 'lb1',
                    listeners: [
                        {
                            name: 'listener1',
                        },
                    ],
                },
                {
                    name: 'lb2',
                    domainName: 'api.example.com',
                    domainZone: zone,
                    listeners: [
                        {
                            name: 'listener2',
                        },
                        {
                            name: 'listener3',
                            protocol: aws_elasticloadbalancingv2_1.ApplicationProtocol.HTTPS,
                            certificate: aws_certificatemanager_1.Certificate.fromCertificateArn(stack, 'Cert', 'helloworld'),
                        },
                    ],
                },
            ],
            targetGroups: [
                {
                    containerPort: 80,
                },
                {
                    containerPort: 90,
                },
            ],
        });
        // THEN
        expect(ecsService.loadBalancer.node.id).toEqual('lb1');
        expect(ecsService.listener.node.id).toEqual('listener1');
        expect(ecsService.targetGroup.node.id).toEqual('ECSTargetGroupweb80Group');
    });
    test('setting vpc and cluster throws error', () => {
        // GIVEN
        const stack = new core_1.Stack();
        const vpc = new aws_ec2_1.Vpc(stack, 'VPC');
        const cluster = new aws_ecs_1.Cluster(stack, 'Cluster', { vpc });
        // WHEN
        expect(() => new lib_1.ApplicationMultipleTargetGroupsEc2Service(stack, 'Service', {
            cluster,
            vpc,
            taskImageOptions: {
                image: aws_ecs_1.ContainerImage.fromRegistry('/aws/aws-example-app'),
            },
        })).toThrow(/You can only specify either vpc or cluster. Alternatively, you can leave both blank/);
    });
    test('creates AWS Cloud Map service for Private DNS namespace', () => {
        // GIVEN
        const stack = new core_1.Stack();
        const vpc = new aws_ec2_1.Vpc(stack, 'MyVpc', {});
        const cluster = new aws_ecs_1.Cluster(stack, 'EcsCluster', { vpc });
        cluster.addAsgCapacityProvider(new aws_ecs_1.AsgCapacityProvider(stack, 'DefaultAutoScalingGroupProvider', {
            autoScalingGroup: new aws_autoscaling_1.AutoScalingGroup(stack, 'DefaultAutoScalingGroup', {
                vpc,
                instanceType: new ec2.InstanceType('t2.micro'),
                machineImage: aws_ec2_1.MachineImage.latestAmazonLinux(),
            }),
        }));
        // WHEN
        cluster.addDefaultCloudMapNamespace({
            name: 'foo.com',
            type: aws_servicediscovery_1.NamespaceType.DNS_PRIVATE,
        });
        new lib_1.ApplicationMultipleTargetGroupsEc2Service(stack, 'Service', {
            cluster,
            taskImageOptions: {
                image: aws_ecs_1.ContainerImage.fromRegistry('hello'),
            },
            cloudMapOptions: {
                name: 'myApp',
            },
            memoryLimitMiB: 512,
        });
        // THEN
        assertions_1.Template.fromStack(stack).hasResourceProperties('AWS::ECS::Service', {
            ServiceRegistries: [
                {
                    ContainerName: 'web',
                    ContainerPort: 80,
                    RegistryArn: {
                        'Fn::GetAtt': [
                            'ServiceCloudmapServiceDE76B29D',
                            'Arn',
                        ],
                    },
                },
            ],
        });
        assertions_1.Template.fromStack(stack).hasResourceProperties('AWS::ServiceDiscovery::Service', {
            DnsConfig: {
                DnsRecords: [
                    {
                        TTL: 60,
                        Type: 'SRV',
                    },
                ],
                NamespaceId: {
                    'Fn::GetAtt': [
                        'EcsClusterDefaultServiceDiscoveryNamespaceB0971B2F',
                        'Id',
                    ],
                },
                RoutingPolicy: 'MULTIVALUE',
            },
            HealthCheckCustomConfig: {
                FailureThreshold: 1,
            },
            Name: 'myApp',
            NamespaceId: {
                'Fn::GetAtt': [
                    'EcsClusterDefaultServiceDiscoveryNamespaceB0971B2F',
                    'Id',
                ],
            },
        });
    });
    test('errors when setting both taskDefinition and taskImageOptions', () => {
        // GIVEN
        const stack = new core_1.Stack();
        const vpc = new aws_ec2_1.Vpc(stack, 'VPC');
        const cluster = new aws_ecs_1.Cluster(stack, 'Cluster', { vpc });
        cluster.addAsgCapacityProvider(new aws_ecs_1.AsgCapacityProvider(stack, 'DefaultAutoScalingGroupProvider', {
            autoScalingGroup: new aws_autoscaling_1.AutoScalingGroup(stack, 'DefaultAutoScalingGroup', {
                vpc,
                instanceType: new ec2.InstanceType('t2.micro'),
                machineImage: aws_ec2_1.MachineImage.latestAmazonLinux(),
            }),
        }));
        const taskDefinition = new aws_ecs_1.Ec2TaskDefinition(stack, 'Ec2TaskDef');
        taskDefinition.addContainer('test', {
            image: aws_ecs_1.ContainerImage.fromRegistry('amazon/amazon-ecs-sample'),
            memoryLimitMiB: 512,
        });
        // THEN
        expect(() => {
            new lib_1.ApplicationMultipleTargetGroupsEc2Service(stack, 'Service', {
                cluster,
                taskImageOptions: {
                    image: aws_ecs_1.ContainerImage.fromRegistry('test'),
                },
                taskDefinition,
            });
        }).toThrow(/You must specify only one of TaskDefinition or TaskImageOptions./);
    });
    test('errors when setting neither taskDefinition nor taskImageOptions', () => {
        // GIVEN
        const stack = new core_1.Stack();
        const vpc = new aws_ec2_1.Vpc(stack, 'VPC');
        const cluster = new aws_ecs_1.Cluster(stack, 'Cluster', { vpc });
        cluster.addAsgCapacityProvider(new aws_ecs_1.AsgCapacityProvider(stack, 'DefaultAutoScalingGroupProvider', {
            autoScalingGroup: new aws_autoscaling_1.AutoScalingGroup(stack, 'DefaultAutoScalingGroup', {
                vpc,
                instanceType: new ec2.InstanceType('t2.micro'),
                machineImage: aws_ec2_1.MachineImage.latestAmazonLinux(),
            }),
        }));
        // THEN
        expect(() => {
            new lib_1.ApplicationMultipleTargetGroupsEc2Service(stack, 'Service', {
                cluster,
            });
        }).toThrow(/You must specify one of: taskDefinition or image/);
    });
    test('errors when setting domainName but not domainZone', () => {
        // GIVEN
        const stack = new core_1.Stack();
        const vpc = new aws_ec2_1.Vpc(stack, 'VPC');
        const cluster = new aws_ecs_1.Cluster(stack, 'Cluster', { vpc });
        cluster.addAsgCapacityProvider(new aws_ecs_1.AsgCapacityProvider(stack, 'DefaultAutoScalingGroupProvider', {
            autoScalingGroup: new aws_autoscaling_1.AutoScalingGroup(stack, 'DefaultAutoScalingGroup', {
                vpc,
                instanceType: new ec2.InstanceType('t2.micro'),
                machineImage: aws_ec2_1.MachineImage.latestAmazonLinux(),
            }),
        }));
        // THEN
        expect(() => {
            new lib_1.ApplicationMultipleTargetGroupsEc2Service(stack, 'Service', {
                cluster,
                taskImageOptions: {
                    image: aws_ecs_1.ContainerImage.fromRegistry('test'),
                },
                loadBalancers: [
                    {
                        name: 'lb1',
                        domainName: 'api.example.com',
                        listeners: [
                            {
                                name: 'listener1',
                            },
                        ],
                    },
                ],
            });
        }).toThrow(/A Route53 hosted domain zone name is required to configure the specified domain name/);
    });
    test('errors when loadBalancers is empty', () => {
        // GIVEN
        const stack = new core_1.Stack();
        const vpc = new aws_ec2_1.Vpc(stack, 'VPC');
        const cluster = new aws_ecs_1.Cluster(stack, 'Cluster', { vpc });
        // THEN
        expect(() => {
            new lib_1.ApplicationMultipleTargetGroupsEc2Service(stack, 'Service', {
                cluster,
                taskImageOptions: {
                    image: aws_ecs_1.ContainerImage.fromRegistry('test'),
                },
                loadBalancers: [],
            });
        }).toThrow(/At least one load balancer must be specified/);
    });
    test('errors when targetGroups is empty', () => {
        // GIVEN
        const stack = new core_1.Stack();
        const vpc = new aws_ec2_1.Vpc(stack, 'VPC');
        const cluster = new aws_ecs_1.Cluster(stack, 'Cluster', { vpc });
        // THEN
        expect(() => {
            new lib_1.ApplicationMultipleTargetGroupsEc2Service(stack, 'Service', {
                cluster,
                taskImageOptions: {
                    image: aws_ecs_1.ContainerImage.fromRegistry('test'),
                },
                targetGroups: [],
            });
        }).toThrow(/At least one target group should be specified/);
    });
    test('errors when no listener specified', () => {
        // GIVEN
        const stack = new core_1.Stack();
        const vpc = new aws_ec2_1.Vpc(stack, 'VPC');
        const cluster = new aws_ecs_1.Cluster(stack, 'Cluster', { vpc });
        // THEN
        expect(() => {
            new lib_1.ApplicationMultipleTargetGroupsEc2Service(stack, 'Service', {
                cluster,
                taskImageOptions: {
                    image: aws_ecs_1.ContainerImage.fromRegistry('test'),
                },
                loadBalancers: [
                    {
                        name: 'lb',
                        listeners: [],
                    },
                ],
            });
        }).toThrow(/At least one listener must be specified/);
    });
    test('errors when setting both HTTP protocol and certificate', () => {
        // GIVEN
        const stack = new core_1.Stack();
        const vpc = new aws_ec2_1.Vpc(stack, 'VPC');
        const cluster = new aws_ecs_1.Cluster(stack, 'Cluster', { vpc });
        // THEN
        expect(() => {
            new lib_1.ApplicationMultipleTargetGroupsEc2Service(stack, 'Service', {
                cluster,
                taskImageOptions: {
                    image: aws_ecs_1.ContainerImage.fromRegistry('test'),
                },
                loadBalancers: [
                    {
                        name: 'lb',
                        listeners: [
                            {
                                name: 'listener',
                                protocol: aws_elasticloadbalancingv2_1.ApplicationProtocol.HTTP,
                                certificate: aws_certificatemanager_1.Certificate.fromCertificateArn(stack, 'Cert', 'helloworld'),
                            },
                        ],
                    },
                ],
            });
        }).toThrow(/The HTTPS protocol must be used when a certificate is given/);
    });
    test('errors when setting HTTPS protocol but not domain name', () => {
        // GIVEN
        const stack = new core_1.Stack();
        const vpc = new aws_ec2_1.Vpc(stack, 'VPC');
        const cluster = new aws_ecs_1.Cluster(stack, 'Cluster', { vpc });
        // THEN
        expect(() => {
            new lib_1.ApplicationMultipleTargetGroupsEc2Service(stack, 'Service', {
                cluster,
                taskImageOptions: {
                    image: aws_ecs_1.ContainerImage.fromRegistry('test'),
                },
                loadBalancers: [
                    {
                        name: 'lb',
                        listeners: [
                            {
                                name: 'listener',
                                protocol: aws_elasticloadbalancingv2_1.ApplicationProtocol.HTTPS,
                            },
                        ],
                    },
                ],
            });
        }).toThrow(/A domain name and zone is required when using the HTTPS protocol/);
    });
    test('errors when listener is not defined but used in creating target groups', () => {
        // GIVEN
        const stack = new core_1.Stack();
        const vpc = new aws_ec2_1.Vpc(stack, 'VPC');
        const cluster = new aws_ecs_1.Cluster(stack, 'Cluster', { vpc });
        // THEN
        expect(() => {
            new lib_1.ApplicationMultipleTargetGroupsEc2Service(stack, 'Service', {
                cluster,
                taskImageOptions: {
                    image: aws_ecs_1.ContainerImage.fromRegistry('test'),
                },
                loadBalancers: [
                    {
                        name: 'lb',
                        listeners: [
                            {
                                name: 'listener1',
                            },
                        ],
                    },
                ],
                targetGroups: [
                    {
                        containerPort: 80,
                        listener: 'listener2',
                    },
                ],
            });
        }).toThrow(/Listener listener2 is not defined. Did you define listener with name listener2?/);
    });
    test('errors if desiredTaskCount is 0', () => {
        // GIVEN
        const stack = new core_1.Stack();
        const vpc = new aws_ec2_1.Vpc(stack, 'VPC');
        const cluster = new aws_ecs_1.Cluster(stack, 'Cluster', { vpc });
        cluster.addAsgCapacityProvider(new aws_ecs_1.AsgCapacityProvider(stack, 'DefaultAutoScalingGroupProvider', {
            autoScalingGroup: new aws_autoscaling_1.AutoScalingGroup(stack, 'DefaultAutoScalingGroup', {
                vpc,
                instanceType: new ec2.InstanceType('t2.micro'),
                machineImage: aws_ec2_1.MachineImage.latestAmazonLinux(),
            }),
        }));
        // THEN
        expect(() => new lib_1.ApplicationMultipleTargetGroupsEc2Service(stack, 'Service', {
            cluster,
            memoryLimitMiB: 1024,
            taskImageOptions: {
                image: aws_ecs_1.ContainerImage.fromRegistry('test'),
            },
            desiredCount: 0,
        })).toThrow(/You must specify a desiredCount greater than 0/);
    });
});
describe('When Network Load Balancer', () => {
    test('test ECS NLB construct with default settings', () => {
        // GIVEN
        const stack = new core_1.Stack();
        const vpc = new aws_ec2_1.Vpc(stack, 'VPC');
        const cluster = new aws_ecs_1.Cluster(stack, 'Cluster', { vpc });
        cluster.addAsgCapacityProvider(new aws_ecs_1.AsgCapacityProvider(stack, 'DefaultAutoScalingGroupProvider', {
            autoScalingGroup: new aws_autoscaling_1.AutoScalingGroup(stack, 'DefaultAutoScalingGroup', {
                vpc,
                instanceType: new ec2.InstanceType('t2.micro'),
                machineImage: aws_ec2_1.MachineImage.latestAmazonLinux(),
            }),
        }));
        // WHEN
        new lib_1.NetworkMultipleTargetGroupsEc2Service(stack, 'Service', {
            cluster,
            memoryLimitMiB: 256,
            taskImageOptions: {
                image: aws_ecs_1.ContainerImage.fromRegistry('test'),
            },
        });
        // THEN - stack contains a load balancer and a service
        assertions_1.Template.fromStack(stack).resourceCountIs('AWS::ElasticLoadBalancingV2::LoadBalancer', 1);
        assertions_1.Template.fromStack(stack).hasResourceProperties('AWS::ECS::Service', {
            LaunchType: 'EC2',
        });
        assertions_1.Template.fromStack(stack).hasResourceProperties('AWS::ECS::TaskDefinition', {
            ContainerDefinitions: [
                {
                    Essential: true,
                    Image: 'test',
                    LogConfiguration: {
                        LogDriver: 'awslogs',
                        Options: {
                            'awslogs-group': {
                                Ref: 'ServiceTaskDefwebLogGroup2A898F61',
                            },
                            'awslogs-stream-prefix': 'Service',
                            'awslogs-region': {
                                Ref: 'AWS::Region',
                            },
                        },
                    },
                    Memory: 256,
                    Name: 'web',
                    PortMappings: [
                        {
                            ContainerPort: 80,
                            HostPort: 0,
                            Protocol: 'tcp',
                        },
                    ],
                },
            ],
            ExecutionRoleArn: {
                'Fn::GetAtt': [
                    'ServiceTaskDefExecutionRole919F7BE3',
                    'Arn',
                ],
            },
            Family: 'ServiceTaskDef79D79521',
            NetworkMode: 'bridge',
            RequiresCompatibilities: [
                'EC2',
            ],
            TaskRoleArn: {
                'Fn::GetAtt': [
                    'ServiceTaskDefTaskRole0CFE2F57',
                    'Arn',
                ],
            },
        });
    });
    test('Assert EnableExecuteCommand is missing if not set', () => {
        // GIVEN
        const stack = new core_1.Stack();
        const vpc = new aws_ec2_1.Vpc(stack, 'VPC');
        const cluster = new aws_ecs_1.Cluster(stack, 'Cluster', { vpc });
        cluster.addAsgCapacityProvider(new aws_ecs_1.AsgCapacityProvider(stack, 'DefaultAutoScalingGroupProvider', {
            autoScalingGroup: new aws_autoscaling_1.AutoScalingGroup(stack, 'DefaultAutoScalingGroup', {
                vpc,
                instanceType: new ec2.InstanceType('t2.micro'),
                machineImage: aws_ec2_1.MachineImage.latestAmazonLinux(),
            }),
        }));
        const zone = new aws_route53_1.PublicHostedZone(stack, 'HostedZone', { zoneName: 'example.com' });
        // WHEN
        new lib_1.NetworkMultipleTargetGroupsEc2Service(stack, 'Service', {
            cluster,
            memoryLimitMiB: 256,
            taskImageOptions: {
                image: aws_ecs_1.ContainerImage.fromRegistry('test'),
                containerName: 'myContainer',
                containerPorts: [80, 90],
                enableLogging: false,
                environment: {
                    TEST_ENVIRONMENT_VARIABLE1: 'test environment variable 1 value',
                    TEST_ENVIRONMENT_VARIABLE2: 'test environment variable 2 value',
                },
                logDriver: new aws_ecs_1.AwsLogDriver({
                    streamPrefix: 'TestStream',
                }),
                family: 'Ec2TaskDef',
                executionRole: new aws_iam_1.Role(stack, 'ExecutionRole', {
                    path: '/',
                    assumedBy: new aws_iam_1.CompositePrincipal(new aws_iam_1.ServicePrincipal('ecs.amazonaws.com'), new aws_iam_1.ServicePrincipal('ecs-tasks.amazonaws.com')),
                }),
                taskRole: new aws_iam_1.Role(stack, 'TaskRole', {
                    assumedBy: new aws_iam_1.ServicePrincipal('ecs-tasks.amazonaws.com'),
                }),
                dockerLabels: { label1: 'labelValue1', label2: 'labelValue2' },
            },
            cpu: 256,
            desiredCount: 3,
            enableECSManagedTags: true,
            enableExecuteCommand: false,
            healthCheckGracePeriod: core_1.Duration.millis(2000),
            loadBalancers: [
                {
                    name: 'lb1',
                    domainName: 'api.example.com',
                    domainZone: zone,
                    publicLoadBalancer: false,
                    listeners: [
                        {
                            name: 'listener1',
                        },
                    ],
                },
                {
                    name: 'lb2',
                    listeners: [
                        {
                            name: 'listener2',
                            port: 81,
                        },
                    ],
                },
            ],
            propagateTags: aws_ecs_1.PropagatedTagSource.SERVICE,
            memoryReservationMiB: 256,
            serviceName: 'myService',
            targetGroups: [
                {
                    containerPort: 80,
                    listener: 'listener1',
                },
                {
                    containerPort: 90,
                    listener: 'listener2',
                },
            ],
            placementStrategies: [aws_ecs_1.PlacementStrategy.spreadAcrossInstances(), aws_ecs_1.PlacementStrategy.packedByCpu(), aws_ecs_1.PlacementStrategy.randomly()],
            placementConstraints: [aws_ecs_1.PlacementConstraint.memberOf('attribute:ecs.instance-type =~ m5a.*')],
        });
        // THEN
        assertions_1.Template.fromStack(stack).hasResourceProperties('AWS::ECS::Service', {
            EnableExecuteCommand: false,
        });
    });
    test('test ECS NLB construct with all settings', () => {
        // GIVEN
        const stack = new core_1.Stack();
        const vpc = new aws_ec2_1.Vpc(stack, 'VPC');
        const cluster = new aws_ecs_1.Cluster(stack, 'Cluster', { vpc });
        cluster.addAsgCapacityProvider(new aws_ecs_1.AsgCapacityProvider(stack, 'DefaultAutoScalingGroupProvider', {
            autoScalingGroup: new aws_autoscaling_1.AutoScalingGroup(stack, 'DefaultAutoScalingGroup', {
                vpc,
                instanceType: new ec2.InstanceType('t2.micro'),
                machineImage: aws_ec2_1.MachineImage.latestAmazonLinux(),
            }),
        }));
        const zone = new aws_route53_1.PublicHostedZone(stack, 'HostedZone', { zoneName: 'example.com' });
        // WHEN
        new lib_1.NetworkMultipleTargetGroupsEc2Service(stack, 'Service', {
            cluster,
            memoryLimitMiB: 256,
            taskImageOptions: {
                image: aws_ecs_1.ContainerImage.fromRegistry('test'),
                containerName: 'myContainer',
                containerPorts: [80, 90],
                enableLogging: false,
                environment: {
                    TEST_ENVIRONMENT_VARIABLE1: 'test environment variable 1 value',
                    TEST_ENVIRONMENT_VARIABLE2: 'test environment variable 2 value',
                },
                logDriver: new aws_ecs_1.AwsLogDriver({
                    streamPrefix: 'TestStream',
                }),
                family: 'Ec2TaskDef',
                executionRole: new aws_iam_1.Role(stack, 'ExecutionRole', {
                    path: '/',
                    assumedBy: new aws_iam_1.CompositePrincipal(new aws_iam_1.ServicePrincipal('ecs.amazonaws.com'), new aws_iam_1.ServicePrincipal('ecs-tasks.amazonaws.com')),
                }),
                taskRole: new aws_iam_1.Role(stack, 'TaskRole', {
                    assumedBy: new aws_iam_1.ServicePrincipal('ecs-tasks.amazonaws.com'),
                }),
                dockerLabels: { label1: 'labelValue1', label2: 'labelValue2' },
            },
            cpu: 256,
            desiredCount: 3,
            enableECSManagedTags: true,
            enableExecuteCommand: true,
            healthCheckGracePeriod: core_1.Duration.millis(2000),
            loadBalancers: [
                {
                    name: 'lb1',
                    domainName: 'api.example.com',
                    domainZone: zone,
                    publicLoadBalancer: false,
                    listeners: [
                        {
                            name: 'listener1',
                        },
                    ],
                },
                {
                    name: 'lb2',
                    listeners: [
                        {
                            name: 'listener2',
                            port: 81,
                        },
                    ],
                },
            ],
            propagateTags: aws_ecs_1.PropagatedTagSource.SERVICE,
            memoryReservationMiB: 256,
            serviceName: 'myService',
            targetGroups: [
                {
                    containerPort: 80,
                    listener: 'listener1',
                },
                {
                    containerPort: 90,
                    listener: 'listener2',
                },
            ],
            placementStrategies: [aws_ecs_1.PlacementStrategy.spreadAcrossInstances(), aws_ecs_1.PlacementStrategy.packedByCpu(), aws_ecs_1.PlacementStrategy.randomly()],
            placementConstraints: [aws_ecs_1.PlacementConstraint.memberOf('attribute:ecs.instance-type =~ m5a.*')],
        });
        // THEN
        assertions_1.Template.fromStack(stack).hasResourceProperties('AWS::ECS::Service', {
            DesiredCount: 3,
            EnableECSManagedTags: true,
            EnableExecuteCommand: true,
            HealthCheckGracePeriodSeconds: 2,
            LaunchType: 'EC2',
            LoadBalancers: [
                {
                    ContainerName: 'myContainer',
                    ContainerPort: 80,
                    TargetGroupArn: {
                        Ref: 'Servicelb1listener1ECSTargetGroupmyContainer80Group43098F8B',
                    },
                },
                {
                    ContainerName: 'myContainer',
                    ContainerPort: 90,
                    TargetGroupArn: {
                        Ref: 'Servicelb2listener2ECSTargetGroupmyContainer90GroupDEB417E4',
                    },
                },
            ],
            PropagateTags: 'SERVICE',
            SchedulingStrategy: 'REPLICA',
            ServiceName: 'myService',
            PlacementConstraints: [{ Type: 'memberOf', Expression: 'attribute:ecs.instance-type =~ m5a.*' }],
            PlacementStrategies: [{ Field: 'instanceId', Type: 'spread' }, { Field: 'CPU', Type: 'binpack' }, { Type: 'random' }],
        });
        assertions_1.Template.fromStack(stack).hasResourceProperties('AWS::ECS::TaskDefinition', {
            ContainerDefinitions: [
                {
                    Cpu: 256,
                    Environment: [
                        {
                            Name: 'TEST_ENVIRONMENT_VARIABLE1',
                            Value: 'test environment variable 1 value',
                        },
                        {
                            Name: 'TEST_ENVIRONMENT_VARIABLE2',
                            Value: 'test environment variable 2 value',
                        },
                    ],
                    Essential: true,
                    Image: 'test',
                    LogConfiguration: {
                        LogDriver: 'awslogs',
                        Options: {
                            'awslogs-group': {
                                Ref: 'ServiceTaskDefmyContainerLogGroup0A87368B',
                            },
                            'awslogs-stream-prefix': 'TestStream',
                            'awslogs-region': {
                                Ref: 'AWS::Region',
                            },
                        },
                    },
                    Memory: 256,
                    MemoryReservation: 256,
                    Name: 'myContainer',
                    PortMappings: [
                        {
                            ContainerPort: 80,
                            HostPort: 0,
                            Protocol: 'tcp',
                        },
                        {
                            ContainerPort: 90,
                            HostPort: 0,
                            Protocol: 'tcp',
                        },
                    ],
                    DockerLabels: {
                        label1: 'labelValue1',
                        label2: 'labelValue2',
                    },
                },
            ],
            ExecutionRoleArn: {
                'Fn::GetAtt': [
                    'ExecutionRole605A040B',
                    'Arn',
                ],
            },
            Family: 'ServiceTaskDef79D79521',
            NetworkMode: 'bridge',
            RequiresCompatibilities: [
                'EC2',
            ],
            TaskRoleArn: {
                'Fn::GetAtt': [
                    'TaskRole30FC0FBB',
                    'Arn',
                ],
            },
        });
    });
    test('EnableExecuteCommand flag generated IAM Permissions', () => {
        const stack = new core_1.Stack();
        const vpc = new aws_ec2_1.Vpc(stack, 'VPC');
        const cluster = new aws_ecs_1.Cluster(stack, 'Cluster', { vpc });
        cluster.addAsgCapacityProvider(new aws_ecs_1.AsgCapacityProvider(stack, 'DefaultAutoScalingGroupProvider', {
            autoScalingGroup: new aws_autoscaling_1.AutoScalingGroup(stack, 'DefaultAutoScalingGroup', {
                vpc,
                instanceType: new ec2.InstanceType('t2.micro'),
                machineImage: aws_ec2_1.MachineImage.latestAmazonLinux(),
            }),
        }));
        const zone = new aws_route53_1.PublicHostedZone(stack, 'HostedZone', { zoneName: 'example.com' });
        // WHEN
        new lib_1.NetworkMultipleTargetGroupsEc2Service(stack, 'Service', {
            cluster,
            memoryLimitMiB: 256,
            taskImageOptions: {
                image: aws_ecs_1.ContainerImage.fromRegistry('test'),
                containerName: 'myContainer',
                containerPorts: [80, 90],
                enableLogging: false,
                environment: {
                    TEST_ENVIRONMENT_VARIABLE1: 'test environment variable 1 value',
                    TEST_ENVIRONMENT_VARIABLE2: 'test environment variable 2 value',
                },
                logDriver: new aws_ecs_1.AwsLogDriver({
                    streamPrefix: 'TestStream',
                }),
                family: 'Ec2TaskDef',
                executionRole: new aws_iam_1.Role(stack, 'ExecutionRole', {
                    path: '/',
                    assumedBy: new aws_iam_1.CompositePrincipal(new aws_iam_1.ServicePrincipal('ecs.amazonaws.com'), new aws_iam_1.ServicePrincipal('ecs-tasks.amazonaws.com')),
                }),
                taskRole: new aws_iam_1.Role(stack, 'TaskRole', {
                    assumedBy: new aws_iam_1.ServicePrincipal('ecs-tasks.amazonaws.com'),
                }),
                dockerLabels: { label1: 'labelValue1', label2: 'labelValue2' },
            },
            cpu: 256,
            desiredCount: 3,
            enableECSManagedTags: true,
            enableExecuteCommand: true,
            healthCheckGracePeriod: core_1.Duration.millis(2000),
            loadBalancers: [
                {
                    name: 'lb1',
                    domainName: 'api.example.com',
                    domainZone: zone,
                    publicLoadBalancer: false,
                    listeners: [
                        {
                            name: 'listener1',
                        },
                    ],
                },
                {
                    name: 'lb2',
                    listeners: [
                        {
                            name: 'listener2',
                            port: 81,
                        },
                    ],
                },
            ],
            propagateTags: aws_ecs_1.PropagatedTagSource.SERVICE,
            memoryReservationMiB: 256,
            serviceName: 'myService',
            targetGroups: [
                {
                    containerPort: 80,
                    listener: 'listener1',
                },
                {
                    containerPort: 90,
                    listener: 'listener2',
                },
            ],
            placementStrategies: [aws_ecs_1.PlacementStrategy.spreadAcrossInstances(), aws_ecs_1.PlacementStrategy.packedByCpu(), aws_ecs_1.PlacementStrategy.randomly()],
            placementConstraints: [aws_ecs_1.PlacementConstraint.memberOf('attribute:ecs.instance-type =~ m5a.*')],
        });
        // ECS Exec
        assertions_1.Template.fromStack(stack).hasResourceProperties('AWS::IAM::Policy', {
            PolicyDocument: {
                Statement: [
                    {
                        Action: [
                            'ssmmessages:CreateControlChannel',
                            'ssmmessages:CreateDataChannel',
                            'ssmmessages:OpenControlChannel',
                            'ssmmessages:OpenDataChannel',
                        ],
                        Effect: 'Allow',
                        Resource: '*',
                    },
                    {
                        Action: 'logs:DescribeLogGroups',
                        Effect: 'Allow',
                        Resource: '*',
                    },
                    {
                        Action: [
                            'logs:CreateLogStream',
                            'logs:DescribeLogStreams',
                            'logs:PutLogEvents',
                        ],
                        Effect: 'Allow',
                        Resource: '*',
                    },
                ],
                Version: '2012-10-17',
            },
            PolicyName: 'TaskRoleDefaultPolicy07FC53DE',
            Roles: [
                {
                    Ref: 'TaskRole30FC0FBB',
                },
            ],
        });
    });
    test('able to pass pre-defined task definition', () => {
        // GIVEN
        const stack = new core_1.Stack();
        const vpc = new aws_ec2_1.Vpc(stack, 'VPC');
        const cluster = new aws_ecs_1.Cluster(stack, 'Cluster', { vpc });
        cluster.addAsgCapacityProvider(new aws_ecs_1.AsgCapacityProvider(stack, 'DefaultAutoScalingGroupProvider', {
            autoScalingGroup: new aws_autoscaling_1.AutoScalingGroup(stack, 'DefaultAutoScalingGroup', {
                vpc,
                instanceType: new ec2.InstanceType('t2.micro'),
                machineImage: aws_ec2_1.MachineImage.latestAmazonLinux(),
            }),
        }));
        const taskDefinition = new aws_ecs_1.Ec2TaskDefinition(stack, 'Ec2TaskDef');
        const container = taskDefinition.addContainer('web', {
            image: aws_ecs_1.ContainerImage.fromRegistry('amazon/amazon-ecs-sample'),
            memoryLimitMiB: 512,
        });
        container.addPortMappings({
            containerPort: 80,
        });
        // WHEN
        new lib_1.NetworkMultipleTargetGroupsEc2Service(stack, 'Service', {
            cluster,
            taskDefinition,
        });
        // THEN
        assertions_1.Template.fromStack(stack).hasResourceProperties('AWS::ECS::TaskDefinition', {
            ContainerDefinitions: [
                {
                    Essential: true,
                    Image: 'amazon/amazon-ecs-sample',
                    Memory: 512,
                    Name: 'web',
                    PortMappings: [
                        {
                            ContainerPort: 80,
                            HostPort: 0,
                            Protocol: 'tcp',
                        },
                    ],
                },
            ],
            Family: 'Ec2TaskDef',
            NetworkMode: 'bridge',
            RequiresCompatibilities: [
                'EC2',
            ],
        });
    });
    test('errors if no essential container in pre-defined task definition', () => {
        // GIVEN
        const stack = new core_1.Stack();
        const vpc = new aws_ec2_1.Vpc(stack, 'VPC');
        const cluster = new aws_ecs_1.Cluster(stack, 'Cluster', { vpc });
        cluster.addAsgCapacityProvider(new aws_ecs_1.AsgCapacityProvider(stack, 'DefaultAutoScalingGroupProvider', {
            autoScalingGroup: new aws_autoscaling_1.AutoScalingGroup(stack, 'DefaultAutoScalingGroup', {
                vpc,
                instanceType: new ec2.InstanceType('t2.micro'),
                machineImage: aws_ec2_1.MachineImage.latestAmazonLinux(),
            }),
        }));
        const taskDefinition = new aws_ecs_1.Ec2TaskDefinition(stack, 'Ec2TaskDef');
        // THEN
        expect(() => {
            new lib_1.NetworkMultipleTargetGroupsEc2Service(stack, 'Service', {
                cluster,
                taskDefinition,
            });
        }).toThrow(/At least one essential container must be specified/);
    });
    test('set default load balancer, listener, target group correctly', () => {
        // GIVEN
        const stack = new core_1.Stack();
        const vpc = new aws_ec2_1.Vpc(stack, 'VPC');
        const zone = new aws_route53_1.PublicHostedZone(stack, 'HostedZone', { zoneName: 'example.com' });
        // WHEN
        const ecsService = new lib_1.NetworkMultipleTargetGroupsEc2Service(stack, 'Service', {
            vpc,
            memoryLimitMiB: 1024,
            taskImageOptions: {
                image: aws_ecs_1.ContainerImage.fromRegistry('test'),
            },
            loadBalancers: [
                {
                    name: 'lb1',
                    listeners: [
                        {
                            name: 'listener1',
                        },
                    ],
                },
                {
                    name: 'lb2',
                    domainName: 'api.example.com',
                    domainZone: zone,
                    listeners: [
                        {
                            name: 'listener2',
                        },
                        {
                            name: 'listener3',
                        },
                    ],
                },
            ],
            targetGroups: [
                {
                    containerPort: 80,
                },
                {
                    containerPort: 90,
                },
            ],
        });
        // THEN
        expect(ecsService.loadBalancer.node.id).toEqual('lb1');
        expect(ecsService.listener.node.id).toEqual('listener1');
        expect(ecsService.targetGroup.node.id).toEqual('ECSTargetGroupweb80Group');
    });
    test('setting vpc and cluster throws error', () => {
        // GIVEN
        const stack = new core_1.Stack();
        const vpc = new aws_ec2_1.Vpc(stack, 'VPC');
        const cluster = new aws_ecs_1.Cluster(stack, 'Cluster', { vpc });
        // WHEN
        expect(() => new lib_1.NetworkMultipleTargetGroupsEc2Service(stack, 'Service', {
            cluster,
            vpc,
            taskImageOptions: {
                image: aws_ecs_1.ContainerImage.fromRegistry('/aws/aws-example-app'),
            },
        })).toThrow(/You can only specify either vpc or cluster. Alternatively, you can leave both blank/);
    });
    test('creates AWS Cloud Map service for Private DNS namespace', () => {
        // GIVEN
        const stack = new core_1.Stack();
        const vpc = new aws_ec2_1.Vpc(stack, 'MyVpc', {});
        const cluster = new aws_ecs_1.Cluster(stack, 'EcsCluster', { vpc });
        cluster.addAsgCapacityProvider(new aws_ecs_1.AsgCapacityProvider(stack, 'DefaultAutoScalingGroupProvider', {
            autoScalingGroup: new aws_autoscaling_1.AutoScalingGroup(stack, 'DefaultAutoScalingGroup', {
                vpc,
                instanceType: new ec2.InstanceType('t2.micro'),
                machineImage: aws_ec2_1.MachineImage.latestAmazonLinux(),
            }),
        }));
        // WHEN
        cluster.addDefaultCloudMapNamespace({
            name: 'foo.com',
            type: aws_servicediscovery_1.NamespaceType.DNS_PRIVATE,
        });
        new lib_1.NetworkMultipleTargetGroupsEc2Service(stack, 'Service', {
            cluster,
            taskImageOptions: {
                image: aws_ecs_1.ContainerImage.fromRegistry('hello'),
            },
            cloudMapOptions: {
                name: 'myApp',
            },
            memoryLimitMiB: 512,
        });
        // THEN
        assertions_1.Template.fromStack(stack).hasResourceProperties('AWS::ECS::Service', {
            ServiceRegistries: [
                {
                    ContainerName: 'web',
                    ContainerPort: 80,
                    RegistryArn: {
                        'Fn::GetAtt': [
                            'ServiceCloudmapServiceDE76B29D',
                            'Arn',
                        ],
                    },
                },
            ],
        });
        assertions_1.Template.fromStack(stack).hasResourceProperties('AWS::ServiceDiscovery::Service', {
            DnsConfig: {
                DnsRecords: [
                    {
                        TTL: 60,
                        Type: 'SRV',
                    },
                ],
                NamespaceId: {
                    'Fn::GetAtt': [
                        'EcsClusterDefaultServiceDiscoveryNamespaceB0971B2F',
                        'Id',
                    ],
                },
                RoutingPolicy: 'MULTIVALUE',
            },
            HealthCheckCustomConfig: {
                FailureThreshold: 1,
            },
            Name: 'myApp',
            NamespaceId: {
                'Fn::GetAtt': [
                    'EcsClusterDefaultServiceDiscoveryNamespaceB0971B2F',
                    'Id',
                ],
            },
        });
    });
    test('errors when setting both taskDefinition and taskImageOptions', () => {
        // GIVEN
        const stack = new core_1.Stack();
        const vpc = new aws_ec2_1.Vpc(stack, 'VPC');
        const cluster = new aws_ecs_1.Cluster(stack, 'Cluster', { vpc });
        cluster.addAsgCapacityProvider(new aws_ecs_1.AsgCapacityProvider(stack, 'DefaultAutoScalingGroupProvider', {
            autoScalingGroup: new aws_autoscaling_1.AutoScalingGroup(stack, 'DefaultAutoScalingGroup', {
                vpc,
                instanceType: new ec2.InstanceType('t2.micro'),
                machineImage: aws_ec2_1.MachineImage.latestAmazonLinux(),
            }),
        }));
        const taskDefinition = new aws_ecs_1.Ec2TaskDefinition(stack, 'Ec2TaskDef');
        taskDefinition.addContainer('test', {
            image: aws_ecs_1.ContainerImage.fromRegistry('amazon/amazon-ecs-sample'),
            memoryLimitMiB: 512,
        });
        // THEN
        expect(() => {
            new lib_1.NetworkMultipleTargetGroupsEc2Service(stack, 'Service', {
                cluster,
                taskImageOptions: {
                    image: aws_ecs_1.ContainerImage.fromRegistry('test'),
                },
                taskDefinition,
            });
        }).toThrow(/You must specify only one of TaskDefinition or TaskImageOptions./);
    });
    test('errors when setting neither taskDefinition nor taskImageOptions', () => {
        // GIVEN
        const stack = new core_1.Stack();
        const vpc = new aws_ec2_1.Vpc(stack, 'VPC');
        const cluster = new aws_ecs_1.Cluster(stack, 'Cluster', { vpc });
        cluster.addAsgCapacityProvider(new aws_ecs_1.AsgCapacityProvider(stack, 'DefaultAutoScalingGroupProvider', {
            autoScalingGroup: new aws_autoscaling_1.AutoScalingGroup(stack, 'DefaultAutoScalingGroup', {
                vpc,
                instanceType: new ec2.InstanceType('t2.micro'),
                machineImage: aws_ec2_1.MachineImage.latestAmazonLinux(),
            }),
        }));
        // THEN
        expect(() => {
            new lib_1.NetworkMultipleTargetGroupsEc2Service(stack, 'Service', {
                cluster,
            });
        }).toThrow(/You must specify one of: taskDefinition or image/);
    });
    test('errors when setting domainName but not domainZone', () => {
        // GIVEN
        const stack = new core_1.Stack();
        const vpc = new aws_ec2_1.Vpc(stack, 'VPC');
        const cluster = new aws_ecs_1.Cluster(stack, 'Cluster', { vpc });
        cluster.addAsgCapacityProvider(new aws_ecs_1.AsgCapacityProvider(stack, 'DefaultAutoScalingGroupProvider', {
            autoScalingGroup: new aws_autoscaling_1.AutoScalingGroup(stack, 'DefaultAutoScalingGroup', {
                vpc,
                instanceType: new ec2.InstanceType('t2.micro'),
                machineImage: aws_ec2_1.MachineImage.latestAmazonLinux(),
            }),
        }));
        // THEN
        expect(() => {
            new lib_1.NetworkMultipleTargetGroupsEc2Service(stack, 'Service', {
                cluster,
                taskImageOptions: {
                    image: aws_ecs_1.ContainerImage.fromRegistry('test'),
                },
                loadBalancers: [
                    {
                        name: 'lb1',
                        domainName: 'api.example.com',
                        listeners: [{
                                name: 'listener1',
                            }],
                    },
                ],
            });
        }).toThrow(/A Route53 hosted domain zone name is required to configure the specified domain name/);
    });
    test('errors when loadBalancers is empty', () => {
        // GIVEN
        const stack = new core_1.Stack();
        const vpc = new aws_ec2_1.Vpc(stack, 'VPC');
        const cluster = new aws_ecs_1.Cluster(stack, 'Cluster', { vpc });
        // THEN
        expect(() => {
            new lib_1.NetworkMultipleTargetGroupsEc2Service(stack, 'Service', {
                cluster,
                taskImageOptions: {
                    image: aws_ecs_1.ContainerImage.fromRegistry('test'),
                },
                loadBalancers: [],
            });
        }).toThrow(/At least one load balancer must be specified/);
    });
    test('errors when targetGroups is empty', () => {
        // GIVEN
        const stack = new core_1.Stack();
        const vpc = new aws_ec2_1.Vpc(stack, 'VPC');
        const cluster = new aws_ecs_1.Cluster(stack, 'Cluster', { vpc });
        // THEN
        expect(() => {
            new lib_1.NetworkMultipleTargetGroupsEc2Service(stack, 'Service', {
                cluster,
                taskImageOptions: {
                    image: aws_ecs_1.ContainerImage.fromRegistry('test'),
                },
                targetGroups: [],
            });
        }).toThrow(/At least one target group should be specified/);
    });
    test('errors when no listener specified', () => {
        // GIVEN
        const stack = new core_1.Stack();
        const vpc = new aws_ec2_1.Vpc(stack, 'VPC');
        const cluster = new aws_ecs_1.Cluster(stack, 'Cluster', { vpc });
        // THEN
        expect(() => {
            new lib_1.NetworkMultipleTargetGroupsEc2Service(stack, 'Service', {
                cluster,
                taskImageOptions: {
                    image: aws_ecs_1.ContainerImage.fromRegistry('test'),
                },
                loadBalancers: [
                    {
                        name: 'lb',
                        listeners: [],
                    },
                ],
            });
        }).toThrow(/At least one listener must be specified/);
    });
    test('errors when listener is not defined but used in creating target groups', () => {
        // GIVEN
        const stack = new core_1.Stack();
        const vpc = new aws_ec2_1.Vpc(stack, 'VPC');
        const cluster = new aws_ecs_1.Cluster(stack, 'Cluster', { vpc });
        // THEN
        expect(() => {
            new lib_1.NetworkMultipleTargetGroupsEc2Service(stack, 'Service', {
                cluster,
                taskImageOptions: {
                    image: aws_ecs_1.ContainerImage.fromRegistry('test'),
                },
                loadBalancers: [
                    {
                        name: 'lb',
                        listeners: [
                            {
                                name: 'listener1',
                            },
                        ],
                    },
                ],
                targetGroups: [
                    {
                        containerPort: 80,
                        listener: 'listener2',
                    },
                ],
            });
        }).toThrow(/Listener listener2 is not defined. Did you define listener with name listener2?/);
    });
    test('errors if desiredTaskCount is 0', () => {
        // GIVEN
        const stack = new core_1.Stack();
        const vpc = new aws_ec2_1.Vpc(stack, 'VPC');
        const cluster = new aws_ecs_1.Cluster(stack, 'Cluster', { vpc });
        cluster.addAsgCapacityProvider(new aws_ecs_1.AsgCapacityProvider(stack, 'DefaultAutoScalingGroupProvider', {
            autoScalingGroup: new aws_autoscaling_1.AutoScalingGroup(stack, 'DefaultAutoScalingGroup', {
                vpc,
                instanceType: new ec2.InstanceType('t2.micro'),
                machineImage: aws_ec2_1.MachineImage.latestAmazonLinux(),
            }),
        }));
        // THEN
        expect(() => new lib_1.NetworkMultipleTargetGroupsEc2Service(stack, 'Service', {
            cluster,
            memoryLimitMiB: 1024,
            taskImageOptions: {
                image: aws_ecs_1.ContainerImage.fromRegistry('test'),
            },
            desiredCount: 0,
        })).toThrow(/You must specify a desiredCount greater than 0/);
    });
});
