// aws-cdk-lib/aws-elasticloadbalancingv2-targets/test/alb-target.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const assertions_1 = require("../../assertions");
const ec2 = require("../../aws-ec2");
const elbv2 = require("../../aws-elasticloadbalancingv2");
const core_1 = require("../../core");
const targets = require("../lib");
test('Can create target groups with alb target', () => {
    // GIVEN
    const stack = new core_1.Stack();
    const vpc = new ec2.Vpc(stack, 'Stack');
    const alb = new elbv2.ApplicationLoadBalancer(stack, 'ALB', { vpc });
    const nlb = new elbv2.NetworkLoadBalancer(stack, 'NLB', { vpc });
    const listener = nlb.addListener('Listener', { port: 80 });
    // WHEN
    listener.addTargets('Targets', {
        targets: [new targets.AlbTarget(alb, 80)],
        port: 80,
    });
    // THEN
    assertions_1.Template.fromStack(stack).hasResourceProperties('AWS::ElasticLoadBalancingV2::TargetGroup', {
        Port: 80,
        Protocol: 'TCP',
        Targets: [
            {
                Id: {
                    Ref: 'ALBAEE750D2',
                },
                Port: 80,
            },
        ],
        TargetType: 'alb',
        VpcId: {
            Ref: 'Stack8A423254',
        },
    });
});
test('Can create target groups with alb arn target', () => {
    // GIVEN
    const stack = new core_1.Stack();
    const vpc = new ec2.Vpc(stack, 'Stack');
    const nlb = new elbv2.NetworkLoadBalancer(stack, 'NLB', { vpc });
    const listener = nlb.addListener('Listener', { port: 80 });
    // WHEN
    listener.addTargets('Targets', {
        targets: [new targets.AlbArnTarget('MOCK_ARN', 80)],
        port: 80,
    });
    // THEN
    assertions_1.Template.fromStack(stack).hasResourceProperties('AWS::ElasticLoadBalancingV2::TargetGroup', {
        Port: 80,
        Protocol: 'TCP',
        Targets: [
            {
                Id: 'MOCK_ARN',
                Port: 80,
            },
        ],
        TargetType: 'alb',
        VpcId: {
            Ref: 'Stack8A423254',
        },
    });
});
