// aws-cdk-lib/aws-elasticloadbalancingv2-targets/test/lambda-target.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const assertions_1 = require("../../assertions");
const ec2 = require("../../aws-ec2");
const elbv2 = require("../../aws-elasticloadbalancingv2");
const lambda = require("../../aws-lambda");
const core_1 = require("../../core");
const targets = require("../lib");
let stack;
let listener;
let fn;
beforeEach(() => {
    stack = new core_1.Stack();
    const vpc = new ec2.Vpc(stack, 'Stack');
    const lb = new elbv2.ApplicationLoadBalancer(stack, 'LB', { vpc });
    listener = lb.addListener('Listener', { port: 80 });
    fn = new lambda.Function(stack, 'Fun', {
        code: lambda.Code.fromInline('foo'),
        runtime: lambda.Runtime.PYTHON_3_9,
        handler: 'index.handler',
    });
});
test('Can create target groups with lambda targets', () => {
    // WHEN
    listener.addTargets('Targets', {
        targets: [new targets.LambdaTarget(fn)],
    });
    // THEN
    assertions_1.Template.fromStack(stack).hasResourceProperties('AWS::ElasticLoadBalancingV2::TargetGroup', {
        TargetType: 'lambda',
        Targets: [
            { Id: { 'Fn::GetAtt': ['FunA2CCED21', 'Arn'] } },
        ],
    });
});
test('Lambda targets create dependency on Invoke permission', () => {
    // WHEN
    listener.addTargets('Targets', {
        targets: [new targets.LambdaTarget(fn)],
    });
    // THEN
    assertions_1.Template.fromStack(stack).hasResource('AWS::ElasticLoadBalancingV2::TargetGroup', (def) => {
        return (def.DependsOn ?? []).includes('FunInvokeServicePrincipalelasticloadbalancingamazonawscomD2CAC0C4');
    });
});
