// aws-cdk-lib/aws-cloudwatch-actions/test/appscaling.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const assertions_1 = require("../../assertions");
const appscaling = require("../../aws-applicationautoscaling");
const cloudwatch = require("../../aws-cloudwatch");
const core_1 = require("../../core");
const actions = require("../lib");
test('can use topic as alarm action', () => {
    // GIVEN
    const stack = new core_1.Stack();
    const scalingTarget = new appscaling.ScalableTarget(stack, 'Target', {
        minCapacity: 1,
        maxCapacity: 100,
        resourceId: 'asdf',
        scalableDimension: 'height',
        serviceNamespace: appscaling.ServiceNamespace.CUSTOM_RESOURCE,
    });
    const action = new appscaling.StepScalingAction(stack, 'Action', {
        scalingTarget,
    });
    const alarm = new cloudwatch.Alarm(stack, 'Alarm', {
        metric: new cloudwatch.Metric({ namespace: 'AWS', metricName: 'Henk' }),
        evaluationPeriods: 3,
        threshold: 100,
    });
    // WHEN
    alarm.addAlarmAction(new actions.ApplicationScalingAction(action));
    // THEN
    assertions_1.Template.fromStack(stack).hasResourceProperties('AWS::CloudWatch::Alarm', {
        AlarmActions: [
            { Ref: 'Action62AD07C0' },
        ],
    });
});
