// aws-cdk-lib/aws-cloudwatch-actions/test/ec2.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const assertions_1 = require("../../assertions");
const cloudwatch = require("../../aws-cloudwatch");
const core_1 = require("../../core");
const actions = require("../lib");
test('can use instance reboot as alarm action', () => {
    // GIVEN
    const stack = new core_1.Stack();
    const alarm = new cloudwatch.Alarm(stack, 'Alarm', {
        metric: new cloudwatch.Metric({
            namespace: 'AWS/EC2',
            metricName: 'StatusCheckFailed',
            dimensionsMap: {
                InstanceId: 'i-03cb889aaaafffeee',
            },
        }),
        evaluationPeriods: 3,
        threshold: 100,
    });
    // WHEN
    alarm.addAlarmAction(new actions.Ec2Action(actions.Ec2InstanceAction.REBOOT));
    // THEN
    assertions_1.Template.fromStack(stack).hasResourceProperties('AWS::CloudWatch::Alarm', {
        AlarmActions: [
            {
                'Fn::Join': [
                    '',
                    [
                        'arn:',
                        {
                            Ref: 'AWS::Partition',
                        },
                        ':automate:',
                        {
                            Ref: 'AWS::Region',
                        },
                        ':ec2:reboot',
                    ],
                ],
            },
        ],
    });
});
