// aws-cdk-lib/aws-cloudwatch-actions/test/scaling.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const assertions_1 = require("../../assertions");
const autoscaling = require("../../aws-autoscaling");
const cloudwatch = require("../../aws-cloudwatch");
const ec2 = require("../../aws-ec2");
const core_1 = require("../../core");
const actions = require("../lib");
test('can use topic as alarm action', () => {
    // GIVEN
    const stack = new core_1.Stack();
    const vpc = new ec2.Vpc(stack, 'Vpc');
    const autoScalingGroup = new autoscaling.AutoScalingGroup(stack, 'ASG', {
        minCapacity: 1,
        maxCapacity: 100,
        instanceType: new ec2.InstanceType('t-1000.macro'),
        machineImage: new ec2.AmazonLinuxImage(),
        vpc,
    });
    const action = new autoscaling.StepScalingAction(stack, 'Action', {
        autoScalingGroup,
    });
    const alarm = new cloudwatch.Alarm(stack, 'Alarm', {
        metric: new cloudwatch.Metric({ namespace: 'AWS', metricName: 'Henk' }),
        evaluationPeriods: 3,
        threshold: 100,
    });
    // WHEN
    alarm.addAlarmAction(new actions.AutoScalingAction(action));
    // THEN
    assertions_1.Template.fromStack(stack).hasResourceProperties('AWS::CloudWatch::Alarm', {
        AlarmActions: [
            { Ref: 'Action62AD07C0' },
        ],
    });
});
