// aws-cdk-lib/aws-cloudwatch-actions/test/ssm.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const assertions_1 = require("../../assertions");
const cloudwatch = require("../../aws-cloudwatch");
const core_1 = require("../../core");
const actions = require("../lib");
test('can use ssm with critical severity and performance category as alarm action', () => {
    // GIVEN
    const stack = new core_1.Stack();
    const alarm = new cloudwatch.Alarm(stack, 'Alarm', {
        metric: new cloudwatch.Metric({
            namespace: 'AWS',
            metricName: 'Test',
        }),
        evaluationPeriods: 3,
        threshold: 100,
    });
    // WHEN
    alarm.addAlarmAction(new actions.SsmAction(actions.OpsItemSeverity.CRITICAL, actions.OpsItemCategory.PERFORMANCE));
    // THEN
    assertions_1.Template.fromStack(stack).hasResourceProperties('AWS::CloudWatch::Alarm', {
        AlarmActions: [
            {
                'Fn::Join': [
                    '',
                    [
                        'arn:',
                        {
                            Ref: 'AWS::Partition',
                        },
                        ':ssm:',
                        {
                            Ref: 'AWS::Region',
                        },
                        ':',
                        {
                            Ref: 'AWS::AccountId',
                        },
                        ':opsitem:1#CATEGORY=Performance',
                    ],
                ],
            },
        ],
    });
});
test('can use ssm with medium severity and no category as alarm action', () => {
    // GIVEN
    const stack = new core_1.Stack();
    const alarm = new cloudwatch.Alarm(stack, 'Alarm', {
        metric: new cloudwatch.Metric({
            namespace: 'AWS',
            metricName: 'Test',
        }),
        evaluationPeriods: 3,
        threshold: 100,
    });
    // WHEN
    alarm.addAlarmAction(new actions.SsmAction(actions.OpsItemSeverity.MEDIUM));
    // THEN
    assertions_1.Template.fromStack(stack).hasResourceProperties('AWS::CloudWatch::Alarm', {
        AlarmActions: [
            {
                'Fn::Join': [
                    '',
                    [
                        'arn:',
                        {
                            Ref: 'AWS::Partition',
                        },
                        ':ssm:',
                        {
                            Ref: 'AWS::Region',
                        },
                        ':',
                        {
                            Ref: 'AWS::AccountId',
                        },
                        ':opsitem:3',
                    ],
                ],
            },
        ],
    });
});
test('can use SSM Incident as alarm action', () => {
    // GIVEN
    const stack = new core_1.Stack();
    const alarm = new cloudwatch.Alarm(stack, 'Alarm', {
        metric: new cloudwatch.Metric({ namespace: 'AWS', metricName: 'Test' }),
        evaluationPeriods: 3,
        threshold: 100,
    });
    // WHEN
    const responsePlanName = 'ResponsePlanName';
    alarm.addAlarmAction(new actions.SsmIncidentAction(responsePlanName));
    // THEN
    assertions_1.Template.fromStack(stack).hasResourceProperties('AWS::CloudWatch::Alarm', {
        AlarmActions: [
            {
                'Fn::Join': [
                    '',
                    [
                        'arn:',
                        {
                            Ref: 'AWS::Partition',
                        },
                        ':ssm-incidents::',
                        {
                            Ref: 'AWS::AccountId',
                        },
                        ':response-plan/ResponsePlanName',
                    ],
                ],
            },
        ],
    });
});
