// aws-cdk-lib/aws-stepfunctions-tasks/test/athena/get-query-execution.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const cdk = require("../../../core");
const get_query_execution_1 = require("../../lib/athena/get-query-execution");
describe('Get Query Execution', () => {
    test('default settings', () => {
        // GIVEN
        const stack = new cdk.Stack();
        // WHEN
        const task = new get_query_execution_1.AthenaGetQueryExecution(stack, 'Query', {
            queryExecutionId: '2da557a1-7283-4c3d-8af9-058348f0bb02',
        });
        // THEN
        expect(stack.resolve(task.toStateJson())).toEqual({
            Type: 'Task',
            Resource: {
                'Fn::Join': [
                    '',
                    [
                        'arn:',
                        {
                            Ref: 'AWS::Partition',
                        },
                        ':states:::athena:getQueryExecution',
                    ],
                ],
            },
            End: true,
            Parameters: {
                QueryExecutionId: '2da557a1-7283-4c3d-8af9-058348f0bb02',
            },
        });
    });
});
