// aws-cdk-lib/aws-rds/test/sql-server/sql-server.instance-engine.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const assertions_1 = require("../../../assertions");
const core = require("../../../core");
const rds = require("../../lib");
describe('sql server instance engine', () => {
    describe('SQL Server instance engine', () => {
        test("has ParameterGroup family ending in '11.0' for major version 11", () => {
            const stack = new core.Stack();
            new rds.ParameterGroup(stack, 'ParameterGroup', {
                engine: rds.DatabaseInstanceEngine.sqlServerWeb({
                    version: rds.SqlServerEngineVersion.VER_11,
                }),
            }).bindToInstance({});
            assertions_1.Template.fromStack(stack).hasResourceProperties('AWS::RDS::DBParameterGroup', {
                Family: 'sqlserver-web-11.0',
            });
        });
        test("has MajorEngineVersion ending in '11.00' for major version 11", () => {
            const stack = new core.Stack();
            new rds.OptionGroup(stack, 'OptionGroup', {
                engine: rds.DatabaseInstanceEngine.sqlServerWeb({
                    version: rds.SqlServerEngineVersion.VER_11,
                }),
                configurations: [
                    {
                        name: 'SQLSERVER_BACKUP_RESTORE',
                        settings: {
                            IAM_ROLE_ARN: 'some-role-arn',
                        },
                    },
                ],
            });
            assertions_1.Template.fromStack(stack).hasResourceProperties('AWS::RDS::OptionGroup', {
                MajorEngineVersion: '11.00',
            });
        });
    });
});
