// aws-cdk-lib/pipelines/test/blueprint/helpers-internal/pipeline-queries.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/* eslint-disable import/no-extraneous-dependencies */
const cdkp = require("../../../lib");
const pipeline_queries_1 = require("../../../lib/helpers-internal/pipeline-queries");
const test_app_1 = require("../../testhelpers/test-app");
let app;
beforeEach(() => {
    app = new test_app_1.TestApp();
});
afterEach(() => {
    app.cleanup();
});
describe('pipeline-queries', () => {
    describe('stackOutputsReferenced', () => {
        let blueprint;
        let stageDeployment;
        let step;
        let queries;
        let stackDeployment;
        let outputName;
        beforeEach(() => {
            blueprint = new Blueprint(app, 'Bp', {
                synth: new cdkp.ShellStep('Synth', {
                    input: cdkp.CodePipelineSource.gitHub('test/test', 'main'),
                    commands: ['build'],
                }),
            });
            const stage = new test_app_1.AppWithOutput(app, 'CrossAccount');
            outputName = 'MyOutput';
            stageDeployment = blueprint.addStage(stage);
            stackDeployment = stageDeployment.stacks[0];
            expect(stackDeployment).not.toBeUndefined();
            step = new cdkp.ShellStep('test', {
                input: cdkp.CodePipelineSource.gitHub('test/test', 'main'),
                commands: ['build'],
                envFromCfnOutputs: {
                    INPUT: stage.theOutput,
                },
            });
            queries = new pipeline_queries_1.PipelineQueries(blueprint);
        });
        const cases = [
            {
                description: 'output referenced in stage pre step',
                additionalSetup: () => stageDeployment.addPre(step),
                expectedResultGetter: () => [outputName],
            },
            {
                description: 'output referenced in stage post step',
                additionalSetup: () => stageDeployment.addPost(step),
                expectedResultGetter: () => [outputName],
            },
            {
                description: 'output referenced in stack pre step',
                additionalSetup: () => stackDeployment.addStackSteps([step], [], []),
                expectedResultGetter: () => [outputName],
            },
            {
                description: 'output referenced in stack changeSet step',
                additionalSetup: () => stackDeployment.addStackSteps([], [step], []),
                expectedResultGetter: () => [outputName],
            },
            {
                description: 'output referenced in stack post step',
                additionalSetup: () => stackDeployment.addStackSteps([], [], [step]),
                expectedResultGetter: () => [outputName],
            },
            {
                description: 'output not referenced',
                additionalSetup: () => { },
                expectedResultGetter: () => [],
            },
        ];
        cases.forEach(testCase => {
            test(testCase.description, () => {
                //WHEN
                testCase.additionalSetup();
                //THEN
                expect(queries.stackOutputsReferenced(stackDeployment)).toEqual(testCase.expectedResultGetter());
            });
        });
    });
});
class Blueprint extends cdkp.PipelineBase {
    doBuildPipeline() {
    }
}
